// jsdom/lib/jsdom/living/dist/attributes.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var DOMException = require("domexception");

var attrGenerated = require("./generated/Attr");

var _require = require("./helpers/strings"),
    asciiLowercase = _require.asciiLowercase;

var _require2 = require("./helpers/namespaces"),
    HTML_NS = _require2.HTML_NS;

var _require3 = require("./helpers/mutation-observers"),
    queueAttributeMutationRecord = _require3.queueAttributeMutationRecord; // The following three are for https://dom.spec.whatwg.org/#concept-element-attribute-has. We don't just have a
// predicate tester since removing that kind of flexibility gives us the potential for better future optimizations.

/* eslint-disable no-restricted-properties */


exports.hasAttribute = function (element, A) {
  return element._attributeList.includes(A);
};

exports.hasAttributeByName = function (element, name) {
  return element._attributesByNameMap.has(name);
};

exports.hasAttributeByNameNS = function (element, namespace, localName) {
  return element._attributeList.some(function (attribute) {
    return attribute._localName === localName && attribute._namespace === namespace;
  });
};

exports.changeAttribute = function (element, attribute, value) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-change
  var _localName = attribute._localName,
      _namespace = attribute._namespace,
      _value = attribute._value;
  queueAttributeMutationRecord(element, _localName, _namespace, _value);
  var oldValue = attribute._value;
  attribute._value = value; // Run jsdom hooks; roughly correspond to spec's "An attribute is set and an attribute is changed."

  element._attrModified(attribute._qualifiedName, value, oldValue);
};

exports.appendAttribute = function (element, attribute) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-append
  var _localName = attribute._localName,
      _namespace = attribute._namespace;
  queueAttributeMutationRecord(element, _localName, _namespace, null);
  var attributeList = element._attributeList;
  attributeList.push(attribute);
  attribute._element = element; // Sync name cache

  var name = attribute._qualifiedName;
  var cache = element._attributesByNameMap;
  var entry = cache.get(name);

  if (!entry) {
    entry = [];
    cache.set(name, entry);
  }

  entry.push(attribute); // Run jsdom hooks; roughly correspond to spec's "An attribute is set and an attribute is added."

  element._attrModified(name, attribute._value, null);
};

exports.removeAttribute = function (element, attribute) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-remove
  var _localName = attribute._localName,
      _namespace = attribute._namespace,
      _value = attribute._value;
  queueAttributeMutationRecord(element, _localName, _namespace, _value);
  var attributeList = element._attributeList;

  for (var i = 0; i < attributeList.length; ++i) {
    if (attributeList[i] === attribute) {
      attributeList.splice(i, 1);
      attribute._element = null; // Sync name cache

      var name = attribute._qualifiedName;
      var cache = element._attributesByNameMap;
      var entry = cache.get(name);
      entry.splice(entry.indexOf(attribute), 1);

      if (entry.length === 0) {
        cache["delete"](name);
      } // Run jsdom hooks; roughly correspond to spec's "An attribute is removed."


      element._attrModified(name, null, attribute._value);

      return;
    }
  }
};

exports.replaceAttribute = function (element, oldAttr, newAttr) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-replace
  var _localName = oldAttr._localName,
      _namespace = oldAttr._namespace,
      _value = oldAttr._value;
  queueAttributeMutationRecord(element, _localName, _namespace, _value);
  var attributeList = element._attributeList;

  for (var i = 0; i < attributeList.length; ++i) {
    if (attributeList[i] === oldAttr) {
      attributeList.splice(i, 1, newAttr);
      oldAttr._element = null;
      newAttr._element = element; // Sync name cache

      var name = newAttr._qualifiedName;
      var cache = element._attributesByNameMap;
      var entry = cache.get(name);

      if (!entry) {
        entry = [];
        cache.set(name, entry);
      }

      entry.splice(entry.indexOf(oldAttr), 1, newAttr); // Run jsdom hooks; roughly correspond to spec's "An attribute is set and an attribute is changed."

      element._attrModified(name, newAttr._value, oldAttr._value);

      return;
    }
  }
};

exports.getAttributeByName = function (element, name) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-get-by-name
  if (element._namespaceURI === HTML_NS && element._ownerDocument._parsingMode === "html") {
    name = asciiLowercase(name);
  }

  var cache = element._attributesByNameMap;
  var entry = cache.get(name);

  if (!entry) {
    return null;
  }

  return entry[0];
};

exports.getAttributeByNameNS = function (element, namespace, localName) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-get-by-namespace
  if (namespace === "") {
    namespace = null;
  }

  var attributeList = element._attributeList;

  for (var i = 0; i < attributeList.length; ++i) {
    var attr = attributeList[i];

    if (attr._namespace === namespace && attr._localName === localName) {
      return attr;
    }
  }

  return null;
}; // Both of the following functions implement https://dom.spec.whatwg.org/#concept-element-attributes-get-value.
// Separated them into two to keep symmetry with other functions.


exports.getAttributeValue = function (element, localName) {
  var attr = exports.getAttributeByNameNS(element, null, localName);

  if (!attr) {
    return "";
  }

  return attr._value;
};

exports.getAttributeValueNS = function (element, namespace, localName) {
  var attr = exports.getAttributeByNameNS(element, namespace, localName);

  if (!attr) {
    return "";
  }

  return attr._value;
};

exports.setAttribute = function (element, attr) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-set
  if (attr._element !== null && attr._element !== element) {
    throw new DOMException("The attribute is in use.", "InUseAttributeError");
  }

  var oldAttr = exports.getAttributeByNameNS(element, attr._namespace, attr._localName);

  if (oldAttr === attr) {
    return attr;
  }

  if (oldAttr !== null) {
    exports.replaceAttribute(element, oldAttr, attr);
  } else {
    exports.appendAttribute(element, attr);
  }

  return oldAttr;
};

exports.setAttributeValue = function (element, localName, value, prefix, namespace) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-set-value
  if (prefix === undefined) {
    prefix = null;
  }

  if (namespace === undefined) {
    namespace = null;
  }

  var attribute = exports.getAttributeByNameNS(element, namespace, localName);

  if (attribute === null) {
    var newAttribute = attrGenerated.createImpl([], {
      namespace: namespace,
      namespacePrefix: prefix,
      localName: localName,
      value: value
    });
    exports.appendAttribute(element, newAttribute);
    return;
  }

  exports.changeAttribute(element, attribute, value);
};

exports.setAnExistingAttributeValue = function (attribute, value) {
  if (attribute._element === null) {
    attribute._value = value;
  }

  exports.changeAttribute(attribute._element, attribute, value);
};

exports.removeAttributeByName = function (element, name) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-remove-by-name
  var attr = exports.getAttributeByName(element, name);

  if (attr !== null) {
    exports.removeAttribute(element, attr);
  }

  return attr;
};

exports.removeAttributeByNameNS = function (element, namespace, localName) {
  // https://dom.spec.whatwg.org/#concept-element-attributes-remove-by-namespace
  var attr = exports.getAttributeByNameNS(element, namespace, localName);

  if (attr !== null) {
    exports.removeAttribute(element, attr);
  }

  return attr;
};

exports.copyAttributeList = function (sourceElement, destElement) {
  // Needed by https://dom.spec.whatwg.org/#concept-node-clone
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = sourceElement._attributeList[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var sourceAttr = _step.value;
      var destAttr = attrGenerated.createImpl([], {
        namespace: sourceAttr._namespace,
        namespacePrefix: sourceAttr._namespacePrefix,
        localName: sourceAttr._localName,
        value: sourceAttr._value
      });
      exports.appendAttribute(destElement, destAttr);
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"] != null) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }
};

exports.attributeListsEqual = function (elementA, elementB) {
  // Needed by https://dom.spec.whatwg.org/#concept-node-equals
  var listA = elementA._attributeList;
  var listB = elementB._attributeList;

  if (listA.length !== listB.length) {
    return false;
  }

  var _loop = function _loop(i) {
    var attrA = listA[i];

    if (!listB.some(function (attrB) {
      return equalsA(attrB);
    })) {
      return {
        v: false
      };
    }

    function equalsA(attrB) {
      return attrA._namespace === attrB._namespace && attrA._localName === attrB._localName && attrA._value === attrB._value;
    }
  };

  for (var i = 0; i < listA.length; ++i) {
    var _ret = _loop(i);

    if (_typeof(_ret) === "object") return _ret.v;
  }

  return true;
};

exports.attributeNames = function (element) {
  // Needed by https://dom.spec.whatwg.org/#dom-element-getattributenames
  return element._attributeList.map(function (a) {
    return a._qualifiedName;
  });
};

exports.hasAttributes = function (element) {
  // Needed by https://dom.spec.whatwg.org/#dom-element-hasattributes
  return element._attributeList.length > 0;
};
