// parcel-bundler/lib/packagers/dist/Packager.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var fs = require('fs');

var _require = require('@parcel/utils'),
    promisify = _require.promisify;

var path = require('path');

var _require2 = require('@parcel/fs'),
    mkdirp = _require2.mkdirp;

var Packager =
/*#__PURE__*/
function () {
  function Packager(bundle, bundler) {
    _classCallCheck(this, Packager);

    this.bundle = bundle;
    this.bundler = bundler;
    this.options = bundler.options;
  }

  _createClass(Packager, [{
    key: "setup",
    value: function setup() {
      var _this = this;

      return (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee() {
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!_this.bundle.name.includes(path.sep)) {
                  _context.next = 3;
                  break;
                }

                _context.next = 3;
                return mkdirp(path.dirname(_this.bundle.name));

              case 3:
                _this.dest = fs.createWriteStream(_this.bundle.name);
                _this.dest.write = promisify(_this.dest.write.bind(_this.dest));
                _this.dest.end = promisify(_this.dest.end.bind(_this.dest));

              case 6:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }))();
    }
  }, {
    key: "write",
    value: function write(string) {
      var _this2 = this;

      return (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee2() {
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return _this2.dest.write(string);

              case 2:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }))();
    }
  }, {
    key: "start",
    value: function start() {
      return (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee3() {
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3);
      }))();
    } // eslint-disable-next-line no-unused-vars

  }, {
    key: "addAsset",
    value: function addAsset(asset) {
      return (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee4() {
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                throw new Error('Must be implemented by subclasses');

              case 1:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4);
      }))();
    }
  }, {
    key: "getSize",
    value: function getSize() {
      return this.dest.bytesWritten;
    }
  }, {
    key: "end",
    value: function end() {
      var _this3 = this;

      return (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee5() {
        return regeneratorRuntime.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return _this3.dest.end();

              case 2:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5);
      }))();
    }
  }], [{
    key: "shouldAddAsset",
    value: function shouldAddAsset() {
      return true;
    }
  }]);

  return Packager;
}();

module.exports = Packager;
