// purgecss/lib/dist/purgecss.dev.js
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function e(e) {
  return e && "object" == _typeof(e) && "default" in e ? e["default"] : e;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var t = require("postcss"),
    s = e(require("postcss-selector-parser")),
    r = require("fs"),
    i = require("util"),
    n = e(require("glob")),
    a = e(require("path"));

var o = {
  css: [],
  content: [],
  defaultExtractor: function defaultExtractor(e) {
    return e.match(/[A-Za-z0-9_-]+/g) || [];
  },
  extractors: [],
  fontFace: !1,
  keyframes: !1,
  rejected: !1,
  stdin: !1,
  stdout: !1,
  variables: !1,
  whitelist: [],
  whitelistPatterns: [],
  whitelistPatternsChildren: []
},
    c = ["*", "::-webkit-scrollbar", "::selection", ":root", "::before", "::after"];

var u = function u(e) {
  _classCallCheck(this, u);

  this.nodes = [], this.isUsed = !1, this.value = e;
};

var l =
/*#__PURE__*/
function () {
  function l() {
    _classCallCheck(this, l);

    this.nodes = new Map(), this.usedVariables = new Set();
  }

  _createClass(l, [{
    key: "addVariable",
    value: function addVariable(e) {
      var t = e.prop;

      if (!this.nodes.has(t)) {
        var _s = new u(e);

        this.nodes.set(t, _s);
      }
    }
  }, {
    key: "addVariableUsage",
    value: function addVariableUsage(e, t) {
      var s = e.prop,
          r = this.nodes.get(s);
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = t[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var _e = _step.value;
          var _t = _e[1];

          if (this.nodes.has(_t)) {
            var _e2 = this.nodes.get(_t);

            null == r || r.nodes.push(_e2);
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }, {
    key: "addVariableUsageInProperties",
    value: function addVariableUsageInProperties(e) {
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = e[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var _t2 = _step2.value;
          var _e3 = _t2[1];
          this.usedVariables.add(_e3);
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
            _iterator2["return"]();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }
  }, {
    key: "setAsUsed",
    value: function setAsUsed(e) {
      var t = [this.nodes.get(e)];

      for (; 0 !== t.length;) {
        var _e4 = t.pop();

        _e4 && !_e4.isUsed && (_e4.isUsed = !0, t.push.apply(t, _toConsumableArray(_e4.nodes)));
      }
    }
  }, {
    key: "removeUnused",
    value: function removeUnused() {
      var _iteratorNormalCompletion3 = true;
      var _didIteratorError3 = false;
      var _iteratorError3 = undefined;

      try {
        for (var _iterator3 = this.usedVariables[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
          var _e5 = _step3.value;
          this.setAsUsed(_e5);
        }
      } catch (err) {
        _didIteratorError3 = true;
        _iteratorError3 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion3 && _iterator3["return"] != null) {
            _iterator3["return"]();
          }
        } finally {
          if (_didIteratorError3) {
            throw _iteratorError3;
          }
        }
      }

      var _iteratorNormalCompletion4 = true;
      var _didIteratorError4 = false;
      var _iteratorError4 = undefined;

      try {
        for (var _iterator4 = this.nodes[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
          var _step4$value = _slicedToArray(_step4.value, 2),
              _e6 = _step4$value[1];

          _e6.isUsed || _e6.value.remove();
        }
      } catch (err) {
        _didIteratorError4 = true;
        _iteratorError4 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion4 && _iterator4["return"] != null) {
            _iterator4["return"]();
          }
        } finally {
          if (_didIteratorError4) {
            throw _iteratorError4;
          }
        }
      }
    }
  }]);

  return l;
}();

function h(e, t) {
  t && t.forEach(e.add, e);
}

var d =
/*#__PURE__*/
function () {
  function d(e) {
    _classCallCheck(this, d);

    this.undetermined = new Set(), this.attrNames = new Set(), this.attrValues = new Set(), this.classes = new Set(), this.ids = new Set(), this.tags = new Set(), this.merge(e);
  }

  _createClass(d, [{
    key: "merge",
    value: function merge(e) {
      return Array.isArray(e) ? h(this.undetermined, e) : e instanceof d ? (h(this.undetermined, e.undetermined), h(this.attrNames, e.attrNames), h(this.attrValues, e.attrValues), h(this.classes, e.classes), h(this.ids, e.ids), h(this.tags, e.tags)) : (h(this.undetermined, e.undetermined), e.attributes && (h(this.attrNames, e.attributes.names), h(this.attrValues, e.attributes.values)), h(this.classes, e.classes), h(this.ids, e.ids), h(this.tags, e.tags)), this;
    }
  }, {
    key: "hasAttrName",
    value: function hasAttrName(e) {
      return this.attrNames.has(e) || this.undetermined.has(e);
    }
  }, {
    key: "someAttrValue",
    value: function someAttrValue(e) {
      var _iteratorNormalCompletion5 = true;
      var _didIteratorError5 = false;
      var _iteratorError5 = undefined;

      try {
        for (var _iterator5 = this.attrValues[Symbol.iterator](), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
          var _t3 = _step5.value;
          if (e(_t3)) return !0;
        }
      } catch (err) {
        _didIteratorError5 = true;
        _iteratorError5 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion5 && _iterator5["return"] != null) {
            _iterator5["return"]();
          }
        } finally {
          if (_didIteratorError5) {
            throw _iteratorError5;
          }
        }
      }

      var _iteratorNormalCompletion6 = true;
      var _didIteratorError6 = false;
      var _iteratorError6 = undefined;

      try {
        for (var _iterator6 = this.undetermined[Symbol.iterator](), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {
          var _t4 = _step6.value;
          if (e(_t4)) return !0;
        }
      } catch (err) {
        _didIteratorError6 = true;
        _iteratorError6 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion6 && _iterator6["return"] != null) {
            _iterator6["return"]();
          }
        } finally {
          if (_didIteratorError6) {
            throw _iteratorError6;
          }
        }
      }

      return !1;
    }
  }, {
    key: "hasAttrPrefix",
    value: function hasAttrPrefix(e) {
      return this.someAttrValue(function (t) {
        return t.startsWith(e);
      });
    }
  }, {
    key: "hasAttrSuffix",
    value: function hasAttrSuffix(e) {
      return this.someAttrValue(function (t) {
        return t.endsWith(e);
      });
    }
  }, {
    key: "hasAttrSubstr",
    value: function hasAttrSubstr(e) {
      return this.someAttrValue(function (t) {
        return t.includes(e);
      });
    }
  }, {
    key: "hasAttrValue",
    value: function hasAttrValue(e) {
      return this.attrValues.has(e) || this.undetermined.has(e);
    }
  }, {
    key: "hasClass",
    value: function hasClass(e) {
      return this.classes.has(e) || this.undetermined.has(e);
    }
  }, {
    key: "hasId",
    value: function hasId(e) {
      return this.ids.has(e) || this.undetermined.has(e);
    }
  }, {
    key: "hasTag",
    value: function hasTag(e) {
      return this.tags.has(e) || this.undetermined.has(e);
    }
  }]);

  return d;
}();

var f = {
  access: i.promisify(r.access),
  readFile: i.promisify(r.readFile)
};

function p() {
  var e,
      t,
      _s2,
      _args = arguments;

  return regeneratorRuntime.async(function p$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          e = _args.length > 0 && _args[0] !== undefined ? _args[0] : "purgecss.config.js";
          _context.prev = 1;
          _s2 = a.join(process.cwd(), e);
          _context.next = 5;
          return regeneratorRuntime.awrap(new Promise(function (e) {
            e(function (e) {
              if (e && e.__esModule) return e;
              var t = {};
              return e && Object.keys(e).forEach(function (s) {
                var r = Object.getOwnPropertyDescriptor(e, s);
                Object.defineProperty(t, s, r.get ? r : {
                  enumerable: !0,
                  get: function get() {
                    return e[s];
                  }
                });
              }), t["default"] = e, t;
            }(require(_s2)));
          }));

        case 5:
          t = _context.sent;
          _context.next = 11;
          break;

        case 8:
          _context.prev = 8;
          _context.t0 = _context["catch"](1);
          throw new Error("Error loading the config file ".concat(_context.t0.message));

        case 11:
          return _context.abrupt("return", _objectSpread({}, o, {}, t));

        case 12:
        case "end":
          return _context.stop();
      }
    }
  }, null, null, [[1, 8]]);
}

function m(e, t) {
  return regeneratorRuntime.async(function m$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.t0 = d;
          _context2.next = 3;
          return regeneratorRuntime.awrap(t(e));

        case 3:
          _context2.t1 = _context2.sent;
          return _context2.abrupt("return", new _context2.t0(_context2.t1));

        case 5:
        case "end":
          return _context2.stop();
      }
    }
  });
}

function g(e, t) {
  switch (t) {
    case "next":
      return e.text.includes("purgecss ignore");

    case "start":
      return e.text.includes("purgecss start ignore");

    case "end":
      return e.text.includes("purgecss end ignore");
  }
}

function v() {
  var t = new d([]);

  for (var _len = arguments.length, e = new Array(_len), _key = 0; _key < _len; _key++) {
    e[_key] = arguments[_key];
  }

  return e.forEach(t.merge, t), t;
}

function y(e) {
  return e.replace(/(^["'])|(["']$)/g, "");
}

function w(e, t) {
  if (!t.hasAttrName(e.attribute)) return !1;
  if (void 0 === e.value) return !0;

  switch (e.operator) {
    case "$=":
      return t.hasAttrSuffix(e.value);

    case "~=":
    case "*=":
      return t.hasAttrSubstr(e.value);

    case "=":
      return t.hasAttrValue(e.value);

    case "|=":
    case "^=":
      return t.hasAttrPrefix(e.value);

    default:
      return !0;
  }
}

function S(e, t) {
  return t.hasId(e.value);
}

function b(e, t) {
  return t.hasTag(e.value);
}

var x =
/*#__PURE__*/
function () {
  function x() {
    _classCallCheck(this, x);

    this.ignore = !1, this.atRules = {
      fontFace: [],
      keyframes: []
    }, this.usedAnimations = new Set(), this.usedFontFaces = new Set(), this.selectorsRemoved = new Set(), this.variablesStructure = new l(), this.options = o;
  }

  _createClass(x, [{
    key: "collectDeclarationsData",
    value: function collectDeclarationsData(e) {
      var t = e.prop,
          s = e.value;

      if (this.options.variables) {
        var _r = function (e, t) {
          var s = [];
          return e.replace(t, function () {
            var t = arguments,
                r = Array.prototype.slice.call(t, 0, -2);
            return r.input = t[t.length - 1], r.index = t[t.length - 2], s.push(r), e;
          }), s;
        }(s, /var\((.+?)[,)]/g);

        t.startsWith("--") ? (this.variablesStructure.addVariable(e), _r.length > 0 && this.variablesStructure.addVariableUsage(e, _r)) : _r.length > 0 && this.variablesStructure.addVariableUsageInProperties(_r);
      }

      if (!this.options.keyframes || "animation" !== t && "animation-name" !== t) {
        if (this.options.fontFace) {
          if ("font-family" === t) {
            var _iteratorNormalCompletion7 = true;
            var _didIteratorError7 = false;
            var _iteratorError7 = undefined;

            try {
              for (var _iterator7 = s.split(",")[Symbol.iterator](), _step7; !(_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done); _iteratorNormalCompletion7 = true) {
                var _e7 = _step7.value;

                var _t5 = y(_e7.trim());

                this.usedFontFaces.add(_t5);
              }
            } catch (err) {
              _didIteratorError7 = true;
              _iteratorError7 = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion7 && _iterator7["return"] != null) {
                  _iterator7["return"]();
                }
              } finally {
                if (_didIteratorError7) {
                  throw _iteratorError7;
                }
              }
            }
          }
        } else ;
      } else {
        var _iteratorNormalCompletion8 = true;
        var _didIteratorError8 = false;
        var _iteratorError8 = undefined;

        try {
          for (var _iterator8 = s.split(/[\s,]+/)[Symbol.iterator](), _step8; !(_iteratorNormalCompletion8 = (_step8 = _iterator8.next()).done); _iteratorNormalCompletion8 = true) {
            var _e8 = _step8.value;
            this.usedAnimations.add(_e8);
          }
        } catch (err) {
          _didIteratorError8 = true;
          _iteratorError8 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion8 && _iterator8["return"] != null) {
              _iterator8["return"]();
            }
          } finally {
            if (_didIteratorError8) {
              throw _iteratorError8;
            }
          }
        }
      }
    }
  }, {
    key: "getFileExtractor",
    value: function getFileExtractor(e, t) {
      var s = t.find(function (t) {
        return t.extensions.find(function (t) {
          return e.endsWith(t);
        });
      });
      return void 0 === s ? this.options.defaultExtractor : s.extractor;
    }
  }, {
    key: "extractSelectorsFromFiles",
    value: function extractSelectorsFromFiles(e, t) {
      var s, _iteratorNormalCompletion9, _didIteratorError9, _iteratorError9, _iterator9, _step9, _i2, _e9, _iteratorNormalCompletion10, _didIteratorError10, _iteratorError10, _iterator10, _step10, _r2, _e10, _i3, _n2;

      return regeneratorRuntime.async(function extractSelectorsFromFiles$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              s = new d([]);
              _iteratorNormalCompletion9 = true;
              _didIteratorError9 = false;
              _iteratorError9 = undefined;
              _context3.prev = 4;
              _iterator9 = e[Symbol.iterator]();

            case 6:
              if (_iteratorNormalCompletion9 = (_step9 = _iterator9.next()).done) {
                _context3.next = 53;
                break;
              }

              _i2 = _step9.value;
              _e9 = [];
              _context3.prev = 9;
              _context3.next = 12;
              return regeneratorRuntime.awrap(f.access(_i2, r.constants.F_OK));

            case 12:
              _e9.push(_i2);

              _context3.next = 18;
              break;

            case 15:
              _context3.prev = 15;
              _context3.t0 = _context3["catch"](9);
              _e9 = n.sync(_i2, {
                nodir: !0
              });

            case 18:
              _iteratorNormalCompletion10 = true;
              _didIteratorError10 = false;
              _iteratorError10 = undefined;
              _context3.prev = 21;
              _iterator10 = _e9[Symbol.iterator]();

            case 23:
              if (_iteratorNormalCompletion10 = (_step10 = _iterator10.next()).done) {
                _context3.next = 36;
                break;
              }

              _r2 = _step10.value;
              _context3.next = 27;
              return regeneratorRuntime.awrap(f.readFile(_r2, "utf-8"));

            case 27:
              _e10 = _context3.sent;
              _i3 = this.getFileExtractor(_r2, t);
              _context3.next = 31;
              return regeneratorRuntime.awrap(m(_e10, _i3));

            case 31:
              _n2 = _context3.sent;
              s.merge(_n2);

            case 33:
              _iteratorNormalCompletion10 = true;
              _context3.next = 23;
              break;

            case 36:
              _context3.next = 42;
              break;

            case 38:
              _context3.prev = 38;
              _context3.t1 = _context3["catch"](21);
              _didIteratorError10 = true;
              _iteratorError10 = _context3.t1;

            case 42:
              _context3.prev = 42;
              _context3.prev = 43;

              if (!_iteratorNormalCompletion10 && _iterator10["return"] != null) {
                _iterator10["return"]();
              }

            case 45:
              _context3.prev = 45;

              if (!_didIteratorError10) {
                _context3.next = 48;
                break;
              }

              throw _iteratorError10;

            case 48:
              return _context3.finish(45);

            case 49:
              return _context3.finish(42);

            case 50:
              _iteratorNormalCompletion9 = true;
              _context3.next = 6;
              break;

            case 53:
              _context3.next = 59;
              break;

            case 55:
              _context3.prev = 55;
              _context3.t2 = _context3["catch"](4);
              _didIteratorError9 = true;
              _iteratorError9 = _context3.t2;

            case 59:
              _context3.prev = 59;
              _context3.prev = 60;

              if (!_iteratorNormalCompletion9 && _iterator9["return"] != null) {
                _iterator9["return"]();
              }

            case 62:
              _context3.prev = 62;

              if (!_didIteratorError9) {
                _context3.next = 65;
                break;
              }

              throw _iteratorError9;

            case 65:
              return _context3.finish(62);

            case 66:
              return _context3.finish(59);

            case 67:
              return _context3.abrupt("return", s);

            case 68:
            case "end":
              return _context3.stop();
          }
        }
      }, null, this, [[4, 55, 59, 67], [9, 15], [21, 38, 42, 50], [43,, 45, 49], [60,, 62, 66]]);
    }
  }, {
    key: "extractSelectorsFromString",
    value: function extractSelectorsFromString(e, t) {
      var s, _iteratorNormalCompletion11, _didIteratorError11, _iteratorError11, _iterator11, _step11, _step11$value, _r3, _i4, _e11, _n3;

      return regeneratorRuntime.async(function extractSelectorsFromString$(_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              s = new d([]);
              _iteratorNormalCompletion11 = true;
              _didIteratorError11 = false;
              _iteratorError11 = undefined;
              _context4.prev = 4;
              _iterator11 = e[Symbol.iterator]();

            case 6:
              if (_iteratorNormalCompletion11 = (_step11 = _iterator11.next()).done) {
                _context4.next = 16;
                break;
              }

              _step11$value = _step11.value, _r3 = _step11$value.raw, _i4 = _step11$value.extension;
              _e11 = this.getFileExtractor(".".concat(_i4), t);
              _context4.next = 11;
              return regeneratorRuntime.awrap(m(_r3, _e11));

            case 11:
              _n3 = _context4.sent;
              s.merge(_n3);

            case 13:
              _iteratorNormalCompletion11 = true;
              _context4.next = 6;
              break;

            case 16:
              _context4.next = 22;
              break;

            case 18:
              _context4.prev = 18;
              _context4.t0 = _context4["catch"](4);
              _didIteratorError11 = true;
              _iteratorError11 = _context4.t0;

            case 22:
              _context4.prev = 22;
              _context4.prev = 23;

              if (!_iteratorNormalCompletion11 && _iterator11["return"] != null) {
                _iterator11["return"]();
              }

            case 25:
              _context4.prev = 25;

              if (!_didIteratorError11) {
                _context4.next = 28;
                break;
              }

              throw _iteratorError11;

            case 28:
              return _context4.finish(25);

            case 29:
              return _context4.finish(22);

            case 30:
              return _context4.abrupt("return", s);

            case 31:
            case "end":
              return _context4.stop();
          }
        }
      }, null, this, [[4, 18, 22, 30], [23,, 25, 29]]);
    }
  }, {
    key: "evaluateAtRule",
    value: function evaluateAtRule(e) {
      if (this.options.keyframes && e.name.endsWith("keyframes")) this.atRules.keyframes.push(e);else if (this.options.fontFace && "font-face" === e.name && e.nodes) {
        var _iteratorNormalCompletion12 = true;
        var _didIteratorError12 = false;
        var _iteratorError12 = undefined;

        try {
          for (var _iterator12 = e.nodes[Symbol.iterator](), _step12; !(_iteratorNormalCompletion12 = (_step12 = _iterator12.next()).done); _iteratorNormalCompletion12 = true) {
            var _t6 = _step12.value;
            "decl" === _t6.type && "font-family" === _t6.prop && this.atRules.fontFace.push({
              name: y(_t6.value),
              node: e
            });
          }
        } catch (err) {
          _didIteratorError12 = true;
          _iteratorError12 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion12 && _iterator12["return"] != null) {
              _iterator12["return"]();
            }
          } finally {
            if (_didIteratorError12) {
              throw _iteratorError12;
            }
          }
        }
      }
    }
  }, {
    key: "evaluateRule",
    value: function evaluateRule(e, t) {
      var _this = this;

      var r, i, _iteratorNormalCompletion13, _didIteratorError13, _iteratorError13, _iterator13, _step13, _t7, n;

      return regeneratorRuntime.async(function evaluateRule$(_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              if (!this.ignore) {
                _context5.next = 2;
                break;
              }

              return _context5.abrupt("return");

            case 2:
              r = e.prev();

              if (!(r && "comment" === r.type && g(r, "next"))) {
                _context5.next = 5;
                break;
              }

              return _context5.abrupt("return", void r.remove());

            case 5:
              if (!(e.parent && "atrule" === e.parent.type && "keyframes" === e.parent.name)) {
                _context5.next = 7;
                break;
              }

              return _context5.abrupt("return");

            case 7:
              if (!("rule" !== e.type)) {
                _context5.next = 9;
                break;
              }

              return _context5.abrupt("return");

            case 9:
              if (!function (e) {
                var t = !1;
                return e.walkComments(function (e) {
                  e && "comment" === e.type && e.text.includes("purgecss ignore current") && (t = !0, e.remove());
                }), t;
              }(e)) {
                _context5.next = 11;
                break;
              }

              return _context5.abrupt("return");

            case 11:
              i = !0;

              if (!(e.selector = s(function (e) {
                e.walk(function (e) {
                  "selector" === e.type && (i = _this.shouldKeepSelector(e, t), i || (_this.options.rejected && _this.selectorsRemoved.add(e.toString()), e.remove()));
                });
              }).processSync(e.selector), i && void 0 !== e.nodes)) {
                _context5.next = 32;
                break;
              }

              _iteratorNormalCompletion13 = true;
              _didIteratorError13 = false;
              _iteratorError13 = undefined;
              _context5.prev = 16;

              for (_iterator13 = e.nodes[Symbol.iterator](); !(_iteratorNormalCompletion13 = (_step13 = _iterator13.next()).done); _iteratorNormalCompletion13 = true) {
                _t7 = _step13.value;
                "decl" === _t7.type && this.collectDeclarationsData(_t7);
              }

              _context5.next = 24;
              break;

            case 20:
              _context5.prev = 20;
              _context5.t0 = _context5["catch"](16);
              _didIteratorError13 = true;
              _iteratorError13 = _context5.t0;

            case 24:
              _context5.prev = 24;
              _context5.prev = 25;

              if (!_iteratorNormalCompletion13 && _iterator13["return"] != null) {
                _iterator13["return"]();
              }

            case 27:
              _context5.prev = 27;

              if (!_didIteratorError13) {
                _context5.next = 30;
                break;
              }

              throw _iteratorError13;

            case 30:
              return _context5.finish(27);

            case 31:
              return _context5.finish(24);

            case 32:
              n = e.parent;
              e.selector || e.remove(), function (e) {
                return !!("rule" === e.type && !e.selector || e.nodes && !e.nodes.length || "atrule" === e.type && (!e.nodes && !e.params || !e.params && e.nodes && !e.nodes.length));
              }(n) && n.remove();

            case 34:
            case "end":
              return _context5.stop();
          }
        }
      }, null, this, [[16, 20, 24, 32], [25,, 27, 31]]);
    }
  }, {
    key: "getPurgedCSS",
    value: function getPurgedCSS(e, s) {
      var r, i, _iteratorNormalCompletion14, _didIteratorError14, _iteratorError14, _iterator14, _step14, _t8, _i6, _i5, _e12, _i7, _n4, _a;

      return regeneratorRuntime.async(function getPurgedCSS$(_context6) {
        while (1) {
          switch (_context6.prev = _context6.next) {
            case 0:
              r = [], i = [];
              _iteratorNormalCompletion14 = true;
              _didIteratorError14 = false;
              _iteratorError14 = undefined;
              _context6.prev = 4;

              for (_iterator14 = e[Symbol.iterator](); !(_iteratorNormalCompletion14 = (_step14 = _iterator14.next()).done); _iteratorNormalCompletion14 = true) {
                _t8 = _step14.value;
                "string" == typeof _t8 ? i.push.apply(i, _toConsumableArray(n.sync(_t8, {
                  nodir: !0
                }))) : i.push(_t8);
              }

              _context6.next = 12;
              break;

            case 8:
              _context6.prev = 8;
              _context6.t0 = _context6["catch"](4);
              _didIteratorError14 = true;
              _iteratorError14 = _context6.t0;

            case 12:
              _context6.prev = 12;
              _context6.prev = 13;

              if (!_iteratorNormalCompletion14 && _iterator14["return"] != null) {
                _iterator14["return"]();
              }

            case 15:
              _context6.prev = 15;

              if (!_didIteratorError14) {
                _context6.next = 18;
                break;
              }

              throw _iteratorError14;

            case 18:
              return _context6.finish(15);

            case 19:
              return _context6.finish(12);

            case 20:
              _i6 = 0, _i5 = i;

            case 21:
              if (!(_i6 < _i5.length)) {
                _context6.next = 43;
                break;
              }

              _e12 = _i5[_i6];

              if (!("string" == typeof _e12)) {
                _context6.next = 34;
                break;
              }

              if (!this.options.stdin) {
                _context6.next = 28;
                break;
              }

              _context6.t2 = _e12;
              _context6.next = 31;
              break;

            case 28:
              _context6.next = 30;
              return regeneratorRuntime.awrap(f.readFile(_e12, "utf-8"));

            case 30:
              _context6.t2 = _context6.sent;

            case 31:
              _context6.t1 = _context6.t2;
              _context6.next = 35;
              break;

            case 34:
              _context6.t1 = _e12.raw;

            case 35:
              _i7 = _context6.t1;
              _n4 = t.parse(_i7);
              this.walkThroughCSS(_n4, s), this.options.fontFace && this.removeUnusedFontFaces(), this.options.keyframes && this.removeUnusedKeyframes(), this.options.variables && this.removeUnusedCSSVariables();
              _a = {
                css: _n4.toString(),
                file: "string" == typeof _e12 ? _e12 : void 0
              };
              "string" == typeof _e12 && (_a.file = _e12), this.options.rejected && (_a.rejected = Array.from(this.selectorsRemoved), this.selectorsRemoved.clear()), r.push(_a);

            case 40:
              _i6++;
              _context6.next = 21;
              break;

            case 43:
              return _context6.abrupt("return", r);

            case 44:
            case "end":
              return _context6.stop();
          }
        }
      }, null, this, [[4, 8, 12, 20], [13,, 15, 19]]);
    }
  }, {
    key: "isSelectorWhitelisted",
    value: function isSelectorWhitelisted(e) {
      return c.includes(e) || this.options.whitelist && this.options.whitelist.some(function (t) {
        return t === e;
      }) || this.options.whitelistPatterns && this.options.whitelistPatterns.some(function (t) {
        return t.test(e);
      });
    }
  }, {
    key: "isSelectorWhitelistedChildren",
    value: function isSelectorWhitelistedChildren(e) {
      return this.options.whitelistPatternsChildren && this.options.whitelistPatternsChildren.some(function (t) {
        return t.test(e);
      });
    }
  }, {
    key: "purge",
    value: function purge(e) {
      var _this$options, t, s, r, i, n, a, c;

      return regeneratorRuntime.async(function purge$(_context7) {
        while (1) {
          switch (_context7.prev = _context7.next) {
            case 0:
              if (!("object" != _typeof(e))) {
                _context7.next = 6;
                break;
              }

              _context7.next = 3;
              return regeneratorRuntime.awrap(p(e));

            case 3:
              _context7.t0 = _context7.sent;
              _context7.next = 7;
              break;

            case 6:
              _context7.t0 = _objectSpread({}, o, {}, e);

            case 7:
              this.options = _context7.t0;
              _this$options = this.options;
              t = _this$options.content;
              s = _this$options.css;
              r = _this$options.extractors;
              i = t.filter(function (e) {
                return "string" == typeof e;
              });
              n = t.filter(function (e) {
                return "object" == _typeof(e);
              });
              _context7.next = 16;
              return regeneratorRuntime.awrap(this.extractSelectorsFromFiles(i, r));

            case 16:
              a = _context7.sent;
              _context7.next = 19;
              return regeneratorRuntime.awrap(this.extractSelectorsFromString(n, r));

            case 19:
              c = _context7.sent;
              return _context7.abrupt("return", this.getPurgedCSS(s, v(a, c)));

            case 21:
            case "end":
              return _context7.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "removeUnusedCSSVariables",
    value: function removeUnusedCSSVariables() {
      this.variablesStructure.removeUnused();
    }
  }, {
    key: "removeUnusedFontFaces",
    value: function removeUnusedFontFaces() {
      var _iteratorNormalCompletion15 = true;
      var _didIteratorError15 = false;
      var _iteratorError15 = undefined;

      try {
        for (var _iterator15 = this.atRules.fontFace[Symbol.iterator](), _step15; !(_iteratorNormalCompletion15 = (_step15 = _iterator15.next()).done); _iteratorNormalCompletion15 = true) {
          var _step15$value = _step15.value,
              _e13 = _step15$value.name,
              _t9 = _step15$value.node;
          this.usedFontFaces.has(_e13) || _t9.remove();
        }
      } catch (err) {
        _didIteratorError15 = true;
        _iteratorError15 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion15 && _iterator15["return"] != null) {
            _iterator15["return"]();
          }
        } finally {
          if (_didIteratorError15) {
            throw _iteratorError15;
          }
        }
      }
    }
  }, {
    key: "removeUnusedKeyframes",
    value: function removeUnusedKeyframes() {
      var _iteratorNormalCompletion16 = true;
      var _didIteratorError16 = false;
      var _iteratorError16 = undefined;

      try {
        for (var _iterator16 = this.atRules.keyframes[Symbol.iterator](), _step16; !(_iteratorNormalCompletion16 = (_step16 = _iterator16.next()).done); _iteratorNormalCompletion16 = true) {
          var _e14 = _step16.value;
          this.usedAnimations.has(_e14.params) || _e14.remove();
        }
      } catch (err) {
        _didIteratorError16 = true;
        _iteratorError16 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion16 && _iterator16["return"] != null) {
            _iterator16["return"]();
          }
        } finally {
          if (_didIteratorError16) {
            throw _iteratorError16;
          }
        }
      }
    }
  }, {
    key: "shouldKeepSelector",
    value: function shouldKeepSelector(e, t) {
      if (function (e) {
        return e.parent && "pseudo" === e.parent.type && e.parent.value.startsWith(":") || !1;
      }(e)) return !0;
      var s = !1;
      var _iteratorNormalCompletion17 = true;
      var _didIteratorError17 = false;
      var _iteratorError17 = undefined;

      try {
        for (var _iterator17 = e.nodes[Symbol.iterator](), _step17; !(_iteratorNormalCompletion17 = (_step17 = _iterator17.next()).done); _iteratorNormalCompletion17 = true) {
          var _i8 = _step17.value;

          var _e15 = "attribute" === _i8.type && _i8.attribute || _i8.value;

          if (_e15 && this.isSelectorWhitelistedChildren(_e15)) return !0;
          if (_e15 && (c.includes(_e15) || this.isSelectorWhitelisted(_e15))) s = !0;else {
            switch (_i8.type) {
              case "attribute":
                s = !!["value", "checked", "selected", "open"].includes(_i8.attribute) || w(_i8, t);
                break;

              case "class":
                r = _i8, s = t.hasClass(r.value);
                break;

              case "id":
                s = S(_i8, t);
                break;

              case "tag":
                s = b(_i8, t);
            }

            if (!s) return !1;
          }
        }
      } catch (err) {
        _didIteratorError17 = true;
        _iteratorError17 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion17 && _iterator17["return"] != null) {
            _iterator17["return"]();
          }
        } finally {
          if (_didIteratorError17) {
            throw _iteratorError17;
          }
        }
      }

      var r;
      return s;
    }
  }, {
    key: "walkThroughCSS",
    value: function walkThroughCSS(e, t) {
      var _this2 = this;

      e.walk(function (e) {
        return "rule" === e.type ? _this2.evaluateRule(e, t) : "atrule" === e.type ? _this2.evaluateAtRule(e) : void ("comment" === e.type && (g(e, "start") ? (_this2.ignore = !0, e.remove()) : g(e, "end") && (_this2.ignore = !1, e.remove())));
      });
    }
  }]);

  return x;
}();

exports.PurgeCSS = x, exports["default"] = x, exports.defaultOptions = o, exports.mergeExtractorSelectors = v, exports.setOptions = p;
