// jsdom/lib/jsdom/living/generated/dist/CharacterData.dev.js
"use strict";

var _Object$definePropert;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var isNode = require("./Node.js").is;

var impl = utils.implSymbol;

var Node = require("./Node.js");

var ChildNode = require("./ChildNode.js");

var NonDocumentTypeChildNode = require("./NonDocumentTypeChildNode.js");

var CharacterData =
/*#__PURE__*/
function (_Node$interface) {
  _inherits(CharacterData, _Node$interface);

  function CharacterData() {
    var _this;

    _classCallCheck(this, CharacterData);

    throw new TypeError("Illegal constructor");
    return _possibleConstructorReturn(_this);
  }

  _createClass(CharacterData, [{
    key: "substringData",
    value: function substringData(offset, count) {
      var _this$impl;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 2) {
        throw new TypeError("Failed to execute 'substringData' on 'CharacterData': 2 arguments required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'substringData' on 'CharacterData': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg = arguments[1];
        _curArg = conversions["unsigned long"](_curArg, {
          context: "Failed to execute 'substringData' on 'CharacterData': parameter 2"
        });
        args.push(_curArg);
      }
      return (_this$impl = this[impl]).substringData.apply(_this$impl, args);
    }
  }, {
    key: "appendData",
    value: function appendData(data) {
      var _this$impl2;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 1) {
        throw new TypeError("Failed to execute 'appendData' on 'CharacterData': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'appendData' on 'CharacterData': parameter 1"
        });
        args.push(curArg);
      }
      return (_this$impl2 = this[impl]).appendData.apply(_this$impl2, args);
    }
  }, {
    key: "insertData",
    value: function insertData(offset, data) {
      var _this$impl3;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 2) {
        throw new TypeError("Failed to execute 'insertData' on 'CharacterData': 2 arguments required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'insertData' on 'CharacterData': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg2 = arguments[1];
        _curArg2 = conversions["DOMString"](_curArg2, {
          context: "Failed to execute 'insertData' on 'CharacterData': parameter 2"
        });
        args.push(_curArg2);
      }
      return (_this$impl3 = this[impl]).insertData.apply(_this$impl3, args);
    }
  }, {
    key: "deleteData",
    value: function deleteData(offset, count) {
      var _this$impl4;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 2) {
        throw new TypeError("Failed to execute 'deleteData' on 'CharacterData': 2 arguments required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'deleteData' on 'CharacterData': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg3 = arguments[1];
        _curArg3 = conversions["unsigned long"](_curArg3, {
          context: "Failed to execute 'deleteData' on 'CharacterData': parameter 2"
        });
        args.push(_curArg3);
      }
      return (_this$impl4 = this[impl]).deleteData.apply(_this$impl4, args);
    }
  }, {
    key: "replaceData",
    value: function replaceData(offset, count, data) {
      var _this$impl5;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 3) {
        throw new TypeError("Failed to execute 'replaceData' on 'CharacterData': 3 arguments required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'replaceData' on 'CharacterData': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg4 = arguments[1];
        _curArg4 = conversions["unsigned long"](_curArg4, {
          context: "Failed to execute 'replaceData' on 'CharacterData': parameter 2"
        });
        args.push(_curArg4);
      }
      {
        var _curArg5 = arguments[2];
        _curArg5 = conversions["DOMString"](_curArg5, {
          context: "Failed to execute 'replaceData' on 'CharacterData': parameter 3"
        });
        args.push(_curArg5);
      }
      return (_this$impl5 = this[impl]).replaceData.apply(_this$impl5, args);
    }
  }, {
    key: "before",
    value: function before() {
      var _this$impl6;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var args = [];

      for (var i = 0; i < arguments.length; i++) {
        var curArg = arguments[i];

        if (isNode(curArg)) {
          curArg = utils.implForWrapper(curArg);
        } else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'before' on 'CharacterData': parameter " + (i + 1)
          });
        }

        args.push(curArg);
      }

      return (_this$impl6 = this[impl]).before.apply(_this$impl6, args);
    }
  }, {
    key: "after",
    value: function after() {
      var _this$impl7;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var args = [];

      for (var i = 0; i < arguments.length; i++) {
        var curArg = arguments[i];

        if (isNode(curArg)) {
          curArg = utils.implForWrapper(curArg);
        } else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'after' on 'CharacterData': parameter " + (i + 1)
          });
        }

        args.push(curArg);
      }

      return (_this$impl7 = this[impl]).after.apply(_this$impl7, args);
    }
  }, {
    key: "replaceWith",
    value: function replaceWith() {
      var _this$impl8;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var args = [];

      for (var i = 0; i < arguments.length; i++) {
        var curArg = arguments[i];

        if (isNode(curArg)) {
          curArg = utils.implForWrapper(curArg);
        } else {
          curArg = conversions["DOMString"](curArg, {
            context: "Failed to execute 'replaceWith' on 'CharacterData': parameter " + (i + 1)
          });
        }

        args.push(curArg);
      }

      return (_this$impl8 = this[impl]).replaceWith.apply(_this$impl8, args);
    }
  }, {
    key: "remove",
    value: function remove() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl].remove();
    }
  }, {
    key: "data",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["data"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'data' property on 'CharacterData': The provided value",
        treatNullAsEmptyString: true
      });
      this[impl]["data"] = V;
    }
  }, {
    key: "length",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["length"];
    }
  }, {
    key: "previousElementSibling",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return utils.tryWrapperForImpl(this[impl]["previousElementSibling"]);
    }
  }, {
    key: "nextElementSibling",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return utils.tryWrapperForImpl(this[impl]["nextElementSibling"]);
    }
  }]);

  return CharacterData;
}(Node["interface"]);

Object.defineProperties(CharacterData.prototype, (_Object$definePropert = {
  substringData: {
    enumerable: true
  },
  appendData: {
    enumerable: true
  },
  insertData: {
    enumerable: true
  },
  deleteData: {
    enumerable: true
  },
  replaceData: {
    enumerable: true
  },
  before: {
    enumerable: true
  },
  after: {
    enumerable: true
  },
  replaceWith: {
    enumerable: true
  },
  remove: {
    enumerable: true
  },
  data: {
    enumerable: true
  },
  length: {
    enumerable: true
  },
  previousElementSibling: {
    enumerable: true
  },
  nextElementSibling: {
    enumerable: true
  }
}, _defineProperty(_Object$definePropert, Symbol.toStringTag, {
  value: "CharacterData",
  configurable: true
}), _defineProperty(_Object$definePropert, Symbol.unscopables, {
  value: {
    before: true,
    after: true,
    replaceWith: true,
    remove: true
  },
  configurable: true
}), _Object$definePropert));
var iface = {
  // When an interface-module that implements this interface as a mixin is loaded, it will append its own `.is()`
  // method into this array. It allows objects that directly implements *those* interfaces to be recognized as
  // implementing this mixin interface.
  _mixedIntoPredicates: [],
  is: function is(obj) {
    if (obj) {
      if (utils.hasOwn(obj, impl) && obj[impl] instanceof Impl.implementation) {
        return true;
      }

      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = module.exports._mixedIntoPredicates[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var isMixedInto = _step.value;

          if (isMixedInto(obj)) {
            return true;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }

    return false;
  },
  isImpl: function isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      }

      var wrapper = utils.wrapperForImpl(obj);
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = module.exports._mixedIntoPredicates[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var isMixedInto = _step2.value;

          if (isMixedInto(wrapper)) {
            return true;
          }
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
            _iterator2["return"]();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }

    return false;
  },
  convert: function convert(obj) {
    var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
        _ref$context = _ref.context,
        context = _ref$context === void 0 ? "The provided value" : _ref$context;

    if (module.exports.is(obj)) {
      return utils.implForWrapper(obj);
    }

    throw new TypeError("".concat(context, " is not of type 'CharacterData'."));
  },
  create: function create(constructorArgs, privateData) {
    var obj = Object.create(CharacterData.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return obj;
  },
  createImpl: function createImpl(constructorArgs, privateData) {
    var obj = Object.create(CharacterData.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return utils.implForWrapper(obj);
  },
  _internalSetup: function _internalSetup(obj) {
    Node._internalSetup(obj);
  },
  setup: function setup(obj, constructorArgs, privateData) {
    if (!privateData) privateData = {};
    privateData.wrapper = obj;

    this._internalSetup(obj);

    Object.defineProperty(obj, impl, {
      value: new Impl.implementation(constructorArgs, privateData),
      configurable: true
    });
    obj[impl][utils.wrapperSymbol] = obj;

    if (Impl.init) {
      Impl.init(obj[impl], privateData);
    }

    return obj;
  },
  "interface": CharacterData,
  expose: {
    Window: {
      CharacterData: CharacterData
    }
  }
}; // iface

module.exports = iface;

ChildNode._mixedIntoPredicates.push(module.exports.is);

NonDocumentTypeChildNode._mixedIntoPredicates.push(module.exports.is);

var Impl = require("../nodes/CharacterData-impl.js");
