// jsdom/lib/jsdom/living/generated/dist/HTMLInputElement.dev.js
"use strict";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var convertSelectionMode = require("./SelectionMode.js").convert;

var convertFileList = require("./FileList.js").convert;

var impl = utils.implSymbol;

var HTMLElement = require("./HTMLElement.js");

var HTMLInputElement =
/*#__PURE__*/
function (_HTMLElement$interfac) {
  _inherits(HTMLInputElement, _HTMLElement$interfac);

  function HTMLInputElement() {
    var _this;

    _classCallCheck(this, HTMLInputElement);

    throw new TypeError("Illegal constructor");
    return _possibleConstructorReturn(_this);
  }

  _createClass(HTMLInputElement, [{
    key: "checkValidity",
    value: function checkValidity() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl].checkValidity();
    }
  }, {
    key: "reportValidity",
    value: function reportValidity() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl].reportValidity();
    }
  }, {
    key: "setCustomValidity",
    value: function setCustomValidity(error) {
      var _this$impl;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 1) {
        throw new TypeError("Failed to execute 'setCustomValidity' on 'HTMLInputElement': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'setCustomValidity' on 'HTMLInputElement': parameter 1"
        });
        args.push(curArg);
      }
      return (_this$impl = this[impl]).setCustomValidity.apply(_this$impl, args);
    }
  }, {
    key: "select",
    value: function select() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl].select();
    }
  }, {
    key: "setRangeText",
    value: function setRangeText(replacement) {
      var _this$impl2;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 1) {
        throw new TypeError("Failed to execute 'setRangeText' on 'HTMLInputElement': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];

      switch (arguments.length) {
        case 1:
          {
            var curArg = arguments[0];
            curArg = conversions["DOMString"](curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 1"
            });
            args.push(curArg);
          }
          break;

        case 2:
          throw new TypeError("Failed to execute 'setRangeText' on 'HTMLInputElement': only " + arguments.length + " arguments present.");
          break;

        case 3:
          {
            var _curArg = arguments[0];
            _curArg = conversions["DOMString"](_curArg, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 1"
            });
            args.push(_curArg);
          }
          {
            var _curArg2 = arguments[1];
            _curArg2 = conversions["unsigned long"](_curArg2, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 2"
            });
            args.push(_curArg2);
          }
          {
            var _curArg3 = arguments[2];
            _curArg3 = conversions["unsigned long"](_curArg3, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 3"
            });
            args.push(_curArg3);
          }
          break;

        default:
          {
            var _curArg4 = arguments[0];
            _curArg4 = conversions["DOMString"](_curArg4, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 1"
            });
            args.push(_curArg4);
          }
          {
            var _curArg5 = arguments[1];
            _curArg5 = conversions["unsigned long"](_curArg5, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 2"
            });
            args.push(_curArg5);
          }
          {
            var _curArg6 = arguments[2];
            _curArg6 = conversions["unsigned long"](_curArg6, {
              context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 3"
            });
            args.push(_curArg6);
          }
          {
            var _curArg7 = arguments[3];

            if (_curArg7 !== undefined) {
              _curArg7 = convertSelectionMode(_curArg7, {
                context: "Failed to execute 'setRangeText' on 'HTMLInputElement': parameter 4"
              });
            } else {
              _curArg7 = "preserve";
            }

            args.push(_curArg7);
          }
      }

      return (_this$impl2 = this[impl]).setRangeText.apply(_this$impl2, args);
    }
  }, {
    key: "setSelectionRange",
    value: function setSelectionRange(start, end) {
      var _this$impl3;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 2) {
        throw new TypeError("Failed to execute 'setSelectionRange' on 'HTMLInputElement': 2 arguments required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["unsigned long"](curArg, {
          context: "Failed to execute 'setSelectionRange' on 'HTMLInputElement': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg8 = arguments[1];
        _curArg8 = conversions["unsigned long"](_curArg8, {
          context: "Failed to execute 'setSelectionRange' on 'HTMLInputElement': parameter 2"
        });
        args.push(_curArg8);
      }
      {
        var _curArg9 = arguments[2];

        if (_curArg9 !== undefined) {
          _curArg9 = conversions["DOMString"](_curArg9, {
            context: "Failed to execute 'setSelectionRange' on 'HTMLInputElement': parameter 3"
          });
        }

        args.push(_curArg9);
      }
      return (_this$impl3 = this[impl]).setSelectionRange.apply(_this$impl3, args);
    }
  }, {
    key: "accept",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "accept");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'accept' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "accept", V);
    }
  }, {
    key: "alt",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "alt");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'alt' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "alt", V);
    }
  }, {
    key: "autocomplete",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "autocomplete");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'autocomplete' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "autocomplete", V);
    }
  }, {
    key: "autofocus",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this.hasAttributeNS(null, "autofocus");
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'autofocus' property on 'HTMLInputElement': The provided value"
      });

      if (V) {
        this.setAttributeNS(null, "autofocus", "");
      } else {
        this.removeAttributeNS(null, "autofocus");
      }
    }
  }, {
    key: "defaultChecked",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this.hasAttributeNS(null, "checked");
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'defaultChecked' property on 'HTMLInputElement': The provided value"
      });

      if (V) {
        this.setAttributeNS(null, "checked", "");
      } else {
        this.removeAttributeNS(null, "checked");
      }
    }
  }, {
    key: "checked",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["checked"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'checked' property on 'HTMLInputElement': The provided value"
      });
      this[impl]["checked"] = V;
    }
  }, {
    key: "dirName",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "dirname");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'dirName' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "dirname", V);
    }
  }, {
    key: "disabled",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this.hasAttributeNS(null, "disabled");
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'disabled' property on 'HTMLInputElement': The provided value"
      });

      if (V) {
        this.setAttributeNS(null, "disabled", "");
      } else {
        this.removeAttributeNS(null, "disabled");
      }
    }
  }, {
    key: "form",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return utils.tryWrapperForImpl(this[impl]["form"]);
    }
  }, {
    key: "files",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return utils.tryWrapperForImpl(this[impl]["files"]);
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (V === null || V === undefined) {
        V = null;
      } else {
        V = convertFileList(V, {
          context: "Failed to set the 'files' property on 'HTMLInputElement': The provided value"
        });
      }

      this[impl]["files"] = V;
    }
  }, {
    key: "formNoValidate",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this.hasAttributeNS(null, "formnovalidate");
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'formNoValidate' property on 'HTMLInputElement': The provided value"
      });

      if (V) {
        this.setAttributeNS(null, "formnovalidate", "");
      } else {
        this.removeAttributeNS(null, "formnovalidate");
      }
    }
  }, {
    key: "formTarget",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "formtarget");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'formTarget' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "formtarget", V);
    }
  }, {
    key: "indeterminate",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["indeterminate"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'indeterminate' property on 'HTMLInputElement': The provided value"
      });
      this[impl]["indeterminate"] = V;
    }
  }, {
    key: "inputMode",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "inputmode");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'inputMode' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "inputmode", V);
    }
  }, {
    key: "list",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return utils.tryWrapperForImpl(this[impl]["list"]);
    }
  }, {
    key: "max",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "max");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'max' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "max", V);
    }
  }, {
    key: "maxLength",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["maxLength"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["long"](V, {
        context: "Failed to set the 'maxLength' property on 'HTMLInputElement': The provided value"
      });
      this[impl]["maxLength"] = V;
    }
  }, {
    key: "min",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "min");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'min' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "min", V);
    }
  }, {
    key: "minLength",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["minLength"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["long"](V, {
        context: "Failed to set the 'minLength' property on 'HTMLInputElement': The provided value"
      });
      this[impl]["minLength"] = V;
    }
  }, {
    key: "multiple",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this.hasAttributeNS(null, "multiple");
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'multiple' property on 'HTMLInputElement': The provided value"
      });

      if (V) {
        this.setAttributeNS(null, "multiple", "");
      } else {
        this.removeAttributeNS(null, "multiple");
      }
    }
  }, {
    key: "name",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "name");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'name' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "name", V);
    }
  }, {
    key: "pattern",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "pattern");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'pattern' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "pattern", V);
    }
  }, {
    key: "placeholder",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "placeholder");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'placeholder' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "placeholder", V);
    }
  }, {
    key: "readOnly",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this.hasAttributeNS(null, "readonly");
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'readOnly' property on 'HTMLInputElement': The provided value"
      });

      if (V) {
        this.setAttributeNS(null, "readonly", "");
      } else {
        this.removeAttributeNS(null, "readonly");
      }
    }
  }, {
    key: "required",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this.hasAttributeNS(null, "required");
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["boolean"](V, {
        context: "Failed to set the 'required' property on 'HTMLInputElement': The provided value"
      });

      if (V) {
        this.setAttributeNS(null, "required", "");
      } else {
        this.removeAttributeNS(null, "required");
      }
    }
  }, {
    key: "size",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["size"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["unsigned long"](V, {
        context: "Failed to set the 'size' property on 'HTMLInputElement': The provided value"
      });
      this[impl]["size"] = V;
    }
  }, {
    key: "src",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["src"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["USVString"](V, {
        context: "Failed to set the 'src' property on 'HTMLInputElement': The provided value"
      });
      this[impl]["src"] = V;
    }
  }, {
    key: "step",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "step");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'step' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "step", V);
    }
  }, {
    key: "type",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["type"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'type' property on 'HTMLInputElement': The provided value"
      });
      this[impl]["type"] = V;
    }
  }, {
    key: "defaultValue",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "value");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'defaultValue' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "value", V);
    }
  }, {
    key: "value",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["value"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'value' property on 'HTMLInputElement': The provided value",
        treatNullAsEmptyString: true
      });
      this[impl]["value"] = V;
    }
  }, {
    key: "willValidate",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["willValidate"];
    }
  }, {
    key: "validity",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return utils.tryWrapperForImpl(this[impl]["validity"]);
    }
  }, {
    key: "validationMessage",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["validationMessage"];
    }
  }, {
    key: "labels",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return utils.tryWrapperForImpl(this[impl]["labels"]);
    }
  }, {
    key: "selectionStart",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["selectionStart"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (V === null || V === undefined) {
        V = null;
      } else {
        V = conversions["unsigned long"](V, {
          context: "Failed to set the 'selectionStart' property on 'HTMLInputElement': The provided value"
        });
      }

      this[impl]["selectionStart"] = V;
    }
  }, {
    key: "selectionEnd",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["selectionEnd"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (V === null || V === undefined) {
        V = null;
      } else {
        V = conversions["unsigned long"](V, {
          context: "Failed to set the 'selectionEnd' property on 'HTMLInputElement': The provided value"
        });
      }

      this[impl]["selectionEnd"] = V;
    }
  }, {
    key: "selectionDirection",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["selectionDirection"];
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (V === null || V === undefined) {
        V = null;
      } else {
        V = conversions["DOMString"](V, {
          context: "Failed to set the 'selectionDirection' property on 'HTMLInputElement': The provided value"
        });
      }

      this[impl]["selectionDirection"] = V;
    }
  }, {
    key: "align",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "align");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'align' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "align", V);
    }
  }, {
    key: "useMap",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      var value = this.getAttributeNS(null, "usemap");
      return value === null ? "" : value;
    },
    set: function set(V) {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      V = conversions["DOMString"](V, {
        context: "Failed to set the 'useMap' property on 'HTMLInputElement': The provided value"
      });
      this.setAttributeNS(null, "usemap", V);
    }
  }]);

  return HTMLInputElement;
}(HTMLElement["interface"]);

Object.defineProperties(HTMLInputElement.prototype, _defineProperty({
  checkValidity: {
    enumerable: true
  },
  reportValidity: {
    enumerable: true
  },
  setCustomValidity: {
    enumerable: true
  },
  select: {
    enumerable: true
  },
  setRangeText: {
    enumerable: true
  },
  setSelectionRange: {
    enumerable: true
  },
  accept: {
    enumerable: true
  },
  alt: {
    enumerable: true
  },
  autocomplete: {
    enumerable: true
  },
  autofocus: {
    enumerable: true
  },
  defaultChecked: {
    enumerable: true
  },
  checked: {
    enumerable: true
  },
  dirName: {
    enumerable: true
  },
  disabled: {
    enumerable: true
  },
  form: {
    enumerable: true
  },
  files: {
    enumerable: true
  },
  formNoValidate: {
    enumerable: true
  },
  formTarget: {
    enumerable: true
  },
  indeterminate: {
    enumerable: true
  },
  inputMode: {
    enumerable: true
  },
  list: {
    enumerable: true
  },
  max: {
    enumerable: true
  },
  maxLength: {
    enumerable: true
  },
  min: {
    enumerable: true
  },
  minLength: {
    enumerable: true
  },
  multiple: {
    enumerable: true
  },
  name: {
    enumerable: true
  },
  pattern: {
    enumerable: true
  },
  placeholder: {
    enumerable: true
  },
  readOnly: {
    enumerable: true
  },
  required: {
    enumerable: true
  },
  size: {
    enumerable: true
  },
  src: {
    enumerable: true
  },
  step: {
    enumerable: true
  },
  type: {
    enumerable: true
  },
  defaultValue: {
    enumerable: true
  },
  value: {
    enumerable: true
  },
  willValidate: {
    enumerable: true
  },
  validity: {
    enumerable: true
  },
  validationMessage: {
    enumerable: true
  },
  labels: {
    enumerable: true
  },
  selectionStart: {
    enumerable: true
  },
  selectionEnd: {
    enumerable: true
  },
  selectionDirection: {
    enumerable: true
  },
  align: {
    enumerable: true
  },
  useMap: {
    enumerable: true
  }
}, Symbol.toStringTag, {
  value: "HTMLInputElement",
  configurable: true
}));
var iface = {
  // When an interface-module that implements this interface as a mixin is loaded, it will append its own `.is()`
  // method into this array. It allows objects that directly implements *those* interfaces to be recognized as
  // implementing this mixin interface.
  _mixedIntoPredicates: [],
  is: function is(obj) {
    if (obj) {
      if (utils.hasOwn(obj, impl) && obj[impl] instanceof Impl.implementation) {
        return true;
      }

      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = module.exports._mixedIntoPredicates[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var isMixedInto = _step.value;

          if (isMixedInto(obj)) {
            return true;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }

    return false;
  },
  isImpl: function isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      }

      var wrapper = utils.wrapperForImpl(obj);
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = module.exports._mixedIntoPredicates[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var isMixedInto = _step2.value;

          if (isMixedInto(wrapper)) {
            return true;
          }
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
            _iterator2["return"]();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }

    return false;
  },
  convert: function convert(obj) {
    var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
        _ref$context = _ref.context,
        context = _ref$context === void 0 ? "The provided value" : _ref$context;

    if (module.exports.is(obj)) {
      return utils.implForWrapper(obj);
    }

    throw new TypeError("".concat(context, " is not of type 'HTMLInputElement'."));
  },
  create: function create(constructorArgs, privateData) {
    var obj = Object.create(HTMLInputElement.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return obj;
  },
  createImpl: function createImpl(constructorArgs, privateData) {
    var obj = Object.create(HTMLInputElement.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return utils.implForWrapper(obj);
  },
  _internalSetup: function _internalSetup(obj) {
    HTMLElement._internalSetup(obj);
  },
  setup: function setup(obj, constructorArgs, privateData) {
    if (!privateData) privateData = {};
    privateData.wrapper = obj;

    this._internalSetup(obj);

    Object.defineProperty(obj, impl, {
      value: new Impl.implementation(constructorArgs, privateData),
      configurable: true
    });
    obj[impl][utils.wrapperSymbol] = obj;

    if (Impl.init) {
      Impl.init(obj[impl], privateData);
    }

    return obj;
  },
  "interface": HTMLInputElement,
  expose: {
    Window: {
      HTMLInputElement: HTMLInputElement
    }
  }
}; // iface

module.exports = iface;

var Impl = require("../nodes/HTMLInputElement-impl.js");
