// jsdom/lib/jsdom/living/generated/dist/MouseEvent.dev.js
"use strict";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var convertMouseEventInit = require("./MouseEventInit.js").convert;

var convertEventTarget = require("./EventTarget.js").convert;

var impl = utils.implSymbol;

var UIEvent = require("./UIEvent.js");

var MouseEvent =
/*#__PURE__*/
function (_UIEvent$interface) {
  _inherits(MouseEvent, _UIEvent$interface);

  function MouseEvent(type) {
    var _this;

    _classCallCheck(this, MouseEvent);

    if (arguments.length < 1) {
      throw new TypeError("Failed to construct 'MouseEvent': 1 argument required, but only " + arguments.length + " present.");
    }

    var args = [];
    {
      var curArg = arguments[0];
      curArg = conversions["DOMString"](curArg, {
        context: "Failed to construct 'MouseEvent': parameter 1"
      });
      args.push(curArg);
    }
    {
      var _curArg = arguments[1];
      _curArg = convertMouseEventInit(_curArg, {
        context: "Failed to construct 'MouseEvent': parameter 2"
      });
      args.push(_curArg);
    }
    return _possibleConstructorReturn(_this, iface.setup(Object.create((this instanceof MouseEvent ? this.constructor : void 0).prototype), args));
  }

  _createClass(MouseEvent, [{
    key: "getModifierState",
    value: function getModifierState(keyArg) {
      var _this$impl;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 1) {
        throw new TypeError("Failed to execute 'getModifierState' on 'MouseEvent': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'getModifierState' on 'MouseEvent': parameter 1"
        });
        args.push(curArg);
      }
      return (_this$impl = this[impl]).getModifierState.apply(_this$impl, args);
    }
  }, {
    key: "initMouseEvent",
    value: function initMouseEvent(typeArg) {
      var _this$impl2;

      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      if (arguments.length < 1) {
        throw new TypeError("Failed to execute 'initMouseEvent' on 'MouseEvent': 1 argument required, but only " + arguments.length + " present.");
      }

      var args = [];
      {
        var curArg = arguments[0];
        curArg = conversions["DOMString"](curArg, {
          context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 1"
        });
        args.push(curArg);
      }
      {
        var _curArg2 = arguments[1];

        if (_curArg2 !== undefined) {
          _curArg2 = conversions["boolean"](_curArg2, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 2"
          });
        } else {
          _curArg2 = false;
        }

        args.push(_curArg2);
      }
      {
        var _curArg3 = arguments[2];

        if (_curArg3 !== undefined) {
          _curArg3 = conversions["boolean"](_curArg3, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 3"
          });
        } else {
          _curArg3 = false;
        }

        args.push(_curArg3);
      }
      {
        var _curArg4 = arguments[3];

        if (_curArg4 !== undefined) {
          if (_curArg4 === null || _curArg4 === undefined) {
            _curArg4 = null;
          } else {
            _curArg4 = utils.tryImplForWrapper(_curArg4);
          }
        } else {
          _curArg4 = null;
        }

        args.push(_curArg4);
      }
      {
        var _curArg5 = arguments[4];

        if (_curArg5 !== undefined) {
          _curArg5 = conversions["long"](_curArg5, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 5"
          });
        } else {
          _curArg5 = 0;
        }

        args.push(_curArg5);
      }
      {
        var _curArg6 = arguments[5];

        if (_curArg6 !== undefined) {
          _curArg6 = conversions["long"](_curArg6, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 6"
          });
        } else {
          _curArg6 = 0;
        }

        args.push(_curArg6);
      }
      {
        var _curArg7 = arguments[6];

        if (_curArg7 !== undefined) {
          _curArg7 = conversions["long"](_curArg7, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 7"
          });
        } else {
          _curArg7 = 0;
        }

        args.push(_curArg7);
      }
      {
        var _curArg8 = arguments[7];

        if (_curArg8 !== undefined) {
          _curArg8 = conversions["long"](_curArg8, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 8"
          });
        } else {
          _curArg8 = 0;
        }

        args.push(_curArg8);
      }
      {
        var _curArg9 = arguments[8];

        if (_curArg9 !== undefined) {
          _curArg9 = conversions["long"](_curArg9, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 9"
          });
        } else {
          _curArg9 = 0;
        }

        args.push(_curArg9);
      }
      {
        var _curArg10 = arguments[9];

        if (_curArg10 !== undefined) {
          _curArg10 = conversions["boolean"](_curArg10, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 10"
          });
        } else {
          _curArg10 = 0;
        }

        args.push(_curArg10);
      }
      {
        var _curArg11 = arguments[10];

        if (_curArg11 !== undefined) {
          _curArg11 = conversions["boolean"](_curArg11, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 11"
          });
        } else {
          _curArg11 = 0;
        }

        args.push(_curArg11);
      }
      {
        var _curArg12 = arguments[11];

        if (_curArg12 !== undefined) {
          _curArg12 = conversions["boolean"](_curArg12, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 12"
          });
        } else {
          _curArg12 = 0;
        }

        args.push(_curArg12);
      }
      {
        var _curArg13 = arguments[12];

        if (_curArg13 !== undefined) {
          _curArg13 = conversions["boolean"](_curArg13, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 13"
          });
        } else {
          _curArg13 = 0;
        }

        args.push(_curArg13);
      }
      {
        var _curArg14 = arguments[13];

        if (_curArg14 !== undefined) {
          _curArg14 = conversions["short"](_curArg14, {
            context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 14"
          });
        } else {
          _curArg14 = 0;
        }

        args.push(_curArg14);
      }
      {
        var _curArg15 = arguments[14];

        if (_curArg15 !== undefined) {
          if (_curArg15 === null || _curArg15 === undefined) {
            _curArg15 = null;
          } else {
            _curArg15 = convertEventTarget(_curArg15, {
              context: "Failed to execute 'initMouseEvent' on 'MouseEvent': parameter 15"
            });
          }
        } else {
          _curArg15 = null;
        }

        args.push(_curArg15);
      }
      return (_this$impl2 = this[impl]).initMouseEvent.apply(_this$impl2, args);
    }
  }, {
    key: "screenX",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["screenX"];
    }
  }, {
    key: "screenY",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["screenY"];
    }
  }, {
    key: "clientX",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["clientX"];
    }
  }, {
    key: "clientY",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["clientY"];
    }
  }, {
    key: "ctrlKey",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["ctrlKey"];
    }
  }, {
    key: "shiftKey",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["shiftKey"];
    }
  }, {
    key: "altKey",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["altKey"];
    }
  }, {
    key: "metaKey",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["metaKey"];
    }
  }, {
    key: "button",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["button"];
    }
  }, {
    key: "buttons",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return this[impl]["buttons"];
    }
  }, {
    key: "relatedTarget",
    get: function get() {
      if (!this || !module.exports.is(this)) {
        throw new TypeError("Illegal invocation");
      }

      return utils.tryWrapperForImpl(this[impl]["relatedTarget"]);
    }
  }]);

  return MouseEvent;
}(UIEvent["interface"]);

Object.defineProperties(MouseEvent.prototype, _defineProperty({
  getModifierState: {
    enumerable: true
  },
  initMouseEvent: {
    enumerable: true
  },
  screenX: {
    enumerable: true
  },
  screenY: {
    enumerable: true
  },
  clientX: {
    enumerable: true
  },
  clientY: {
    enumerable: true
  },
  ctrlKey: {
    enumerable: true
  },
  shiftKey: {
    enumerable: true
  },
  altKey: {
    enumerable: true
  },
  metaKey: {
    enumerable: true
  },
  button: {
    enumerable: true
  },
  buttons: {
    enumerable: true
  },
  relatedTarget: {
    enumerable: true
  }
}, Symbol.toStringTag, {
  value: "MouseEvent",
  configurable: true
}));
var iface = {
  // When an interface-module that implements this interface as a mixin is loaded, it will append its own `.is()`
  // method into this array. It allows objects that directly implements *those* interfaces to be recognized as
  // implementing this mixin interface.
  _mixedIntoPredicates: [],
  is: function is(obj) {
    if (obj) {
      if (utils.hasOwn(obj, impl) && obj[impl] instanceof Impl.implementation) {
        return true;
      }

      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = module.exports._mixedIntoPredicates[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var isMixedInto = _step.value;

          if (isMixedInto(obj)) {
            return true;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }

    return false;
  },
  isImpl: function isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      }

      var wrapper = utils.wrapperForImpl(obj);
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = module.exports._mixedIntoPredicates[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var isMixedInto = _step2.value;

          if (isMixedInto(wrapper)) {
            return true;
          }
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
            _iterator2["return"]();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }

    return false;
  },
  convert: function convert(obj) {
    var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
        _ref$context = _ref.context,
        context = _ref$context === void 0 ? "The provided value" : _ref$context;

    if (module.exports.is(obj)) {
      return utils.implForWrapper(obj);
    }

    throw new TypeError("".concat(context, " is not of type 'MouseEvent'."));
  },
  create: function create(constructorArgs, privateData) {
    var obj = Object.create(MouseEvent.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return obj;
  },
  createImpl: function createImpl(constructorArgs, privateData) {
    var obj = Object.create(MouseEvent.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return utils.implForWrapper(obj);
  },
  _internalSetup: function _internalSetup(obj) {
    UIEvent._internalSetup(obj);
  },
  setup: function setup(obj, constructorArgs, privateData) {
    if (!privateData) privateData = {};
    privateData.wrapper = obj;

    this._internalSetup(obj);

    Object.defineProperty(obj, impl, {
      value: new Impl.implementation(constructorArgs, privateData),
      configurable: true
    });
    obj[impl][utils.wrapperSymbol] = obj;

    if (Impl.init) {
      Impl.init(obj[impl], privateData);
    }

    return obj;
  },
  "interface": MouseEvent,
  expose: {
    Window: {
      MouseEvent: MouseEvent
    }
  }
}; // iface

module.exports = iface;

var Impl = require("../events/MouseEvent-impl.js");
