// jsdom/lib/jsdom/living/generated/dist/WebSocket.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var conversions = require("webidl-conversions");

var utils = require("./utils.js");

var isBlob = require("./Blob.js").is;

var convertBlob = require("./Blob.js").convert;

var BinaryType = require("./BinaryType.js");

var impl = utils.implSymbol;

var EventTarget = require("./EventTarget.js");

module.exports = {
  createInterface: function createInterface() {
    var _Object$definePropert;

    var defaultPrivateData = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var WebSocket =
    /*#__PURE__*/
    function (_EventTarget$interfac) {
      _inherits(WebSocket, _EventTarget$interfac);

      function WebSocket(url) {
        var _this;

        _classCallCheck(this, WebSocket);

        if (arguments.length < 1) {
          throw new TypeError("Failed to construct 'WebSocket': 1 argument required, but only " + arguments.length + " present.");
        }

        var args = [];
        {
          var curArg = arguments[0];
          curArg = conversions["USVString"](curArg, {
            context: "Failed to construct 'WebSocket': parameter 1"
          });
          args.push(curArg);
        }
        {
          var _curArg = arguments[1];

          if (_curArg !== undefined) {
            if (utils.isObject(_curArg)) {
              if (_curArg[Symbol.iterator] !== undefined) {
                if (!utils.isObject(_curArg)) {
                  throw new TypeError("Failed to construct 'WebSocket': parameter 2" + " sequence" + " is not an iterable object.");
                } else {
                  var V = [];
                  var tmp = _curArg;
                  var _iteratorNormalCompletion = true;
                  var _didIteratorError = false;
                  var _iteratorError = undefined;

                  try {
                    for (var _iterator = tmp[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                      var nextItem = _step.value;
                      nextItem = conversions["DOMString"](nextItem, {
                        context: "Failed to construct 'WebSocket': parameter 2" + " sequence" + "'s element"
                      });
                      V.push(nextItem);
                    }
                  } catch (err) {
                    _didIteratorError = true;
                    _iteratorError = err;
                  } finally {
                    try {
                      if (!_iteratorNormalCompletion && _iterator["return"] != null) {
                        _iterator["return"]();
                      }
                    } finally {
                      if (_didIteratorError) {
                        throw _iteratorError;
                      }
                    }
                  }

                  _curArg = V;
                }
              } else {}
            } else {
              _curArg = conversions["DOMString"](_curArg, {
                context: "Failed to construct 'WebSocket': parameter 2"
              });
            }
          } else {
            _curArg = [];
          }

          args.push(_curArg);
        }
        return _possibleConstructorReturn(_this, iface.setup(Object.create((this instanceof WebSocket ? this.constructor : void 0).prototype), args));
      }

      _createClass(WebSocket, [{
        key: "close",
        value: function close() {
          var _this$impl;

          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          var args = [];
          {
            var curArg = arguments[0];

            if (curArg !== undefined) {
              curArg = conversions["unsigned short"](curArg, {
                context: "Failed to execute 'close' on 'WebSocket': parameter 1",
                clamp: true
              });
            }

            args.push(curArg);
          }
          {
            var _curArg2 = arguments[1];

            if (_curArg2 !== undefined) {
              _curArg2 = conversions["USVString"](_curArg2, {
                context: "Failed to execute 'close' on 'WebSocket': parameter 2"
              });
            }

            args.push(_curArg2);
          }
          return (_this$impl = this[impl]).close.apply(_this$impl, args);
        }
      }, {
        key: "send",
        value: function send(data) {
          var _this$impl2;

          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          if (arguments.length < 1) {
            throw new TypeError("Failed to execute 'send' on 'WebSocket': 1 argument required, but only " + arguments.length + " present.");
          }

          var args = [];
          {
            var curArg = arguments[0];

            if (isBlob(curArg)) {
              {
                var _curArg3 = arguments[0];
                _curArg3 = convertBlob(_curArg3, {
                  context: "Failed to execute 'send' on 'WebSocket': parameter 1"
                });
                args.push(_curArg3);
              }
            } else if (curArg instanceof ArrayBuffer || typeof SharedArrayBuffer !== "undefined" && curArg instanceof SharedArrayBuffer) {
              {
                var _curArg4 = arguments[0];
                _curArg4 = conversions["ArrayBuffer"](_curArg4, {
                  context: "Failed to execute 'send' on 'WebSocket': parameter 1"
                });
                args.push(_curArg4);
              }
            } else if (ArrayBuffer.isView(curArg)) {
              {
                var _curArg5 = arguments[0];

                if (ArrayBuffer.isView(_curArg5)) {} else {
                  throw new TypeError("Failed to execute 'send' on 'WebSocket': parameter 1" + " is not of any supported type.");
                }

                args.push(_curArg5);
              }
            } else {
              {
                var _curArg6 = arguments[0];
                _curArg6 = conversions["USVString"](_curArg6, {
                  context: "Failed to execute 'send' on 'WebSocket': parameter 1"
                });
                args.push(_curArg6);
              }
            }
          }
          return (_this$impl2 = this[impl]).send.apply(_this$impl2, args);
        }
      }, {
        key: "url",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return this[impl]["url"];
        }
      }, {
        key: "readyState",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return this[impl]["readyState"];
        }
      }, {
        key: "bufferedAmount",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return this[impl]["bufferedAmount"];
        }
      }, {
        key: "onopen",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return utils.tryWrapperForImpl(this[impl]["onopen"]);
        },
        set: function set(V) {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          V = utils.tryImplForWrapper(V);
          this[impl]["onopen"] = V;
        }
      }, {
        key: "onerror",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return utils.tryWrapperForImpl(this[impl]["onerror"]);
        },
        set: function set(V) {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          V = utils.tryImplForWrapper(V);
          this[impl]["onerror"] = V;
        }
      }, {
        key: "onclose",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return utils.tryWrapperForImpl(this[impl]["onclose"]);
        },
        set: function set(V) {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          V = utils.tryImplForWrapper(V);
          this[impl]["onclose"] = V;
        }
      }, {
        key: "extensions",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return this[impl]["extensions"];
        }
      }, {
        key: "protocol",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return this[impl]["protocol"];
        }
      }, {
        key: "onmessage",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return utils.tryWrapperForImpl(this[impl]["onmessage"]);
        },
        set: function set(V) {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          V = utils.tryImplForWrapper(V);
          this[impl]["onmessage"] = V;
        }
      }, {
        key: "binaryType",
        get: function get() {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          return utils.tryWrapperForImpl(this[impl]["binaryType"]);
        },
        set: function set(V) {
          if (!this || !module.exports.is(this)) {
            throw new TypeError("Illegal invocation");
          }

          V = "".concat(V);

          if (!BinaryType.enumerationValues.has(V)) {
            return;
          }

          this[impl]["binaryType"] = V;
        }
      }]);

      return WebSocket;
    }(EventTarget["interface"]);

    Object.defineProperties(WebSocket.prototype, (_Object$definePropert = {
      close: {
        enumerable: true
      },
      send: {
        enumerable: true
      },
      url: {
        enumerable: true
      },
      readyState: {
        enumerable: true
      },
      bufferedAmount: {
        enumerable: true
      },
      onopen: {
        enumerable: true
      },
      onerror: {
        enumerable: true
      },
      onclose: {
        enumerable: true
      },
      extensions: {
        enumerable: true
      },
      protocol: {
        enumerable: true
      },
      onmessage: {
        enumerable: true
      },
      binaryType: {
        enumerable: true
      }
    }, _defineProperty(_Object$definePropert, Symbol.toStringTag, {
      value: "WebSocket",
      configurable: true
    }), _defineProperty(_Object$definePropert, "CONNECTING", {
      value: 0,
      enumerable: true
    }), _defineProperty(_Object$definePropert, "OPEN", {
      value: 1,
      enumerable: true
    }), _defineProperty(_Object$definePropert, "CLOSING", {
      value: 2,
      enumerable: true
    }), _defineProperty(_Object$definePropert, "CLOSED", {
      value: 3,
      enumerable: true
    }), _Object$definePropert));
    Object.defineProperties(WebSocket, {
      CONNECTING: {
        value: 0,
        enumerable: true
      },
      OPEN: {
        value: 1,
        enumerable: true
      },
      CLOSING: {
        value: 2,
        enumerable: true
      },
      CLOSED: {
        value: 3,
        enumerable: true
      }
    });
    var iface = {
      create: function create(constructorArgs, privateData) {
        var obj = Object.create(WebSocket.prototype);
        obj = this.setup(obj, constructorArgs, privateData);
        return obj;
      },
      createImpl: function createImpl(constructorArgs, privateData) {
        var obj = Object.create(WebSocket.prototype);
        obj = this.setup(obj, constructorArgs, privateData);
        return utils.implForWrapper(obj);
      },
      _internalSetup: function _internalSetup(obj) {
        EventTarget._internalSetup(obj);
      },
      setup: function setup(obj, constructorArgs, privateData) {
        if (!privateData) privateData = {};

        for (var prop in defaultPrivateData) {
          if (!(prop in privateData)) {
            privateData[prop] = defaultPrivateData[prop];
          }
        }

        privateData.wrapper = obj;

        this._internalSetup(obj);

        Object.defineProperty(obj, impl, {
          value: new Impl.implementation(constructorArgs, privateData),
          configurable: true
        });
        obj[impl][utils.wrapperSymbol] = obj;

        if (Impl.init) {
          Impl.init(obj[impl], privateData);
        }

        return obj;
      },
      "interface": WebSocket,
      expose: {
        Window: {
          WebSocket: WebSocket
        },
        Worker: {
          WebSocket: WebSocket
        }
      }
    }; // iface

    return iface;
  },
  // createInterface
  // When an interface-module that implements this interface as a mixin is loaded, it will append its own `.is()`
  // method into this array. It allows objects that directly implements *those* interfaces to be recognized as
  // implementing this mixin interface.
  _mixedIntoPredicates: [],
  is: function is(obj) {
    if (obj) {
      if (utils.hasOwn(obj, impl) && obj[impl] instanceof Impl.implementation) {
        return true;
      }

      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = module.exports._mixedIntoPredicates[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          var isMixedInto = _step2.value;

          if (isMixedInto(obj)) {
            return true;
          }
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
            _iterator2["return"]();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }

    return false;
  },
  isImpl: function isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      }

      var wrapper = utils.wrapperForImpl(obj);
      var _iteratorNormalCompletion3 = true;
      var _didIteratorError3 = false;
      var _iteratorError3 = undefined;

      try {
        for (var _iterator3 = module.exports._mixedIntoPredicates[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
          var isMixedInto = _step3.value;

          if (isMixedInto(wrapper)) {
            return true;
          }
        }
      } catch (err) {
        _didIteratorError3 = true;
        _iteratorError3 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion3 && _iterator3["return"] != null) {
            _iterator3["return"]();
          }
        } finally {
          if (_didIteratorError3) {
            throw _iteratorError3;
          }
        }
      }
    }

    return false;
  },
  convert: function convert(obj) {
    var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
        _ref$context = _ref.context,
        context = _ref$context === void 0 ? "The provided value" : _ref$context;

    if (module.exports.is(obj)) {
      return utils.implForWrapper(obj);
    }

    throw new TypeError("".concat(context, " is not of type 'WebSocket'."));
  }
}; // module.exports

var Impl = require("../websockets/WebSocket-impl.js");
