// parcel-bundler/src/assets/dist/CSSAsset.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _toArray(arr) { return _arrayWithHoles(arr) || _iterableToArray(arr) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var Asset = require('../Asset');

var postcss = require('postcss');

var valueParser = require('postcss-value-parser');

var postcssTransform = require('../transforms/postcss');

var CssSyntaxError = require('postcss/lib/css-syntax-error');

var SourceMap = require('../SourceMap');

var loadSourceMap = require('../utils/loadSourceMap');

var path = require('path');

var urlJoin = require('../utils/urlJoin');

var isURL = require('../utils/is-url');

var URL_RE = /url\s*\("?(?![a-z]+:)/;
var IMPORT_RE = /@import/;
var COMPOSES_RE = /composes:.+from\s*("|').*("|')\s*;?/;
var FROM_IMPORT_RE = /.+from\s*(?:"|')(.*)(?:"|')\s*;?/;
var PROTOCOL_RE = /^[a-z]+:/;

var CSSAsset =
/*#__PURE__*/
function (_Asset) {
  _inherits(CSSAsset, _Asset);

  function CSSAsset(name, options) {
    var _this;

    _classCallCheck(this, CSSAsset);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(CSSAsset).call(this, name, options));
    _this.type = 'css';
    _this.previousSourceMap = _this.options.rendition ? _this.options.rendition.map : null;
    return _this;
  }

  _createClass(CSSAsset, [{
    key: "mightHaveDependencies",
    value: function mightHaveDependencies() {
      return !/\.css$/.test(this.name) || IMPORT_RE.test(this.contents) || COMPOSES_RE.test(this.contents) || URL_RE.test(this.contents);
    }
  }, {
    key: "parse",
    value: function parse(code) {
      var root = postcss.parse(code, {
        from: this.name
      });
      return new CSSAst(code, root);
    }
  }, {
    key: "collectDependencies",
    value: function collectDependencies() {
      var _this2 = this;

      this.ast.root.walkAtRules('import', function (rule) {
        var params = valueParser(rule.params);

        var _params$nodes = _toArray(params.nodes),
            name = _params$nodes[0],
            media = _params$nodes.slice(1);

        var dep;

        if (name.type === 'function' && name.value === 'url' && name.nodes.length) {
          name = name.nodes[0];
        }

        dep = name.value;

        if (!dep) {
          throw new Error('Could not find import name for ' + rule);
        }

        if (PROTOCOL_RE.test(dep)) {
          return;
        } // If this came from an inline <style> tag, don't inline the imported file. Replace with the correct URL instead.
        // TODO: run CSSPackager on inline style tags.


        var inlineHTML = _this2.options.rendition && _this2.options.rendition.inlineHTML;

        if (inlineHTML) {
          name.value = _this2.addURLDependency(dep, {
            loc: rule.source.start
          });
          rule.params = params.toString();
        } else {
          media = valueParser.stringify(media).trim();

          _this2.addDependency(dep, {
            media: media,
            loc: rule.source.start
          });

          rule.remove();
        }

        _this2.ast.dirty = true;
      });
      this.ast.root.walkDecls(function (decl) {
        if (URL_RE.test(decl.value)) {
          var parsed = valueParser(decl.value);
          var dirty = false;
          parsed.walk(function (node) {
            if (node.type === 'function' && node.value === 'url' && node.nodes.length) {
              var url = _this2.addURLDependency(node.nodes[0].value, {
                loc: decl.source.start
              });

              if (!isURL(url)) {
                url = urlJoin(_this2.options.publicURL, url);
              }

              dirty = node.nodes[0].value !== url;
              node.nodes[0].value = url;
            }
          });

          if (dirty) {
            decl.value = parsed.toString();
            _this2.ast.dirty = true;
          }
        }

        if (decl.prop === 'composes' && FROM_IMPORT_RE.test(decl.value)) {
          var _parsed = valueParser(decl.value);

          _parsed.walk(function (node) {
            if (node.type === 'string') {
              var _FROM_IMPORT_RE$exec = FROM_IMPORT_RE.exec(decl.value),
                  _FROM_IMPORT_RE$exec2 = _slicedToArray(_FROM_IMPORT_RE$exec, 2),
                  importPath = _FROM_IMPORT_RE$exec2[1];

              _this2.addURLDependency(importPath, {
                dynamic: false,
                loc: decl.source.start
              });
            }
          });
        }
      });
    }
  }, {
    key: "pretransform",
    value: function pretransform() {
      return regeneratorRuntime.async(function pretransform$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (!(this.options.sourceMaps && !this.previousSourceMap)) {
                _context.next = 4;
                break;
              }

              _context.next = 3;
              return regeneratorRuntime.awrap(loadSourceMap(this));

            case 3:
              this.previousSourceMap = _context.sent;

            case 4:
            case "end":
              return _context.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "transform",
    value: function transform() {
      return regeneratorRuntime.async(function transform$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return regeneratorRuntime.awrap(postcssTransform(this));

            case 2:
            case "end":
              return _context2.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "getCSSAst",
    value: function getCSSAst() {
      // Converts the ast to a CSS ast if needed, so we can apply postcss transforms.
      if (!(this.ast instanceof CSSAst)) {
        this.ast = CSSAsset.prototype.parse.call(this, this.ast.render(this.name));
      }

      return this.ast.root;
    }
  }, {
    key: "generate",
    value: function generate() {
      var _this3 = this;

      var css, result, js;
      return regeneratorRuntime.async(function generate$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              if (this.ast) {
                result = this.ast.render(this.name);
                css = result.css;
                if (result.map) this.sourceMap = result.map;
              } else {
                css = this.contents;
              }

              js = '';

              if (this.options.hmr) {
                this.addDependency('_css_loader');
                js = "\n        var reloadCSS = require('_css_loader');\n        module.hot.dispose(reloadCSS);\n        module.hot.accept(reloadCSS);\n      ";
              }

              if (this.cssModules) {
                js += 'module.exports = ' + JSON.stringify(this.cssModules, null, 2) + ';';
              }

              if (!this.options.sourceMaps) {
                _context3.next = 23;
                break;
              }

              if (!this.sourceMap) {
                _context3.next = 9;
                break;
              }

              _context3.next = 8;
              return regeneratorRuntime.awrap(new SourceMap().addMap(this.sourceMap));

            case 8:
              this.sourceMap = _context3.sent;

            case 9:
              if (!this.previousSourceMap) {
                _context3.next = 22;
                break;
              }

              this.previousSourceMap.sources = this.previousSourceMap.sources.map(function (v) {
                return path.join(path.dirname(_this3.relativeName), _this3.previousSourceMap.sourceRoot || '', v);
              });

              if (!this.sourceMap) {
                _context3.next = 17;
                break;
              }

              _context3.next = 14;
              return regeneratorRuntime.awrap(new SourceMap().extendSourceMap(this.previousSourceMap, this.sourceMap));

            case 14:
              this.sourceMap = _context3.sent;
              _context3.next = 20;
              break;

            case 17:
              _context3.next = 19;
              return regeneratorRuntime.awrap(new SourceMap().addMap(this.previousSourceMap));

            case 19:
              this.sourceMap = _context3.sent;

            case 20:
              _context3.next = 23;
              break;

            case 22:
              if (!this.sourceMap) {
                this.sourceMap = new SourceMap().generateEmptyMap(this.relativeName, css);
              }

            case 23:
              return _context3.abrupt("return", [{
                type: 'css',
                value: css,
                cssModules: this.cssModules,
                map: this.sourceMap
              }, {
                type: 'js',
                value: js,
                hasDependencies: false
              }]);

            case 24:
            case "end":
              return _context3.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "generateErrorMessage",
    value: function generateErrorMessage(err) {
      // Wrap the error in a CssSyntaxError if needed so we can generate a code frame
      if (err.loc && !err.showSourceCode) {
        err = new CssSyntaxError(err.message, err.loc.line, err.loc.column, this.contents);
      }

      err.message = err.reason || err.message;
      err.loc = {
        line: err.line,
        column: err.column
      };

      if (err.showSourceCode) {
        err.codeFrame = err.showSourceCode();
        err.highlightedCodeFrame = err.showSourceCode(true);
      }

      return err;
    }
  }]);

  return CSSAsset;
}(Asset);

var CSSAst =
/*#__PURE__*/
function () {
  function CSSAst(css, root) {
    _classCallCheck(this, CSSAst);

    this.css = css;
    this.root = root;
    this.dirty = false;
  }

  _createClass(CSSAst, [{
    key: "render",
    value: function render(name) {
      if (this.dirty) {
        var _this$root$toResult = this.root.toResult({
          to: name,
          map: {
            inline: false,
            annotation: false,
            sourcesContent: true
          }
        }),
            css = _this$root$toResult.css,
            map = _this$root$toResult.map;

        this.css = css;
        return {
          css: this.css,
          map: map ? map.toJSON() : null
        };
      }

      return {
        css: this.css
      };
    }
  }]);

  return CSSAst;
}();

module.exports = CSSAsset;
