// parcel-bundler/src/assets/dist/GraphqlAsset.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var Asset = require('../Asset');

var localRequire = require('../utils/localRequire');

var Resolver = require('../Resolver');

var fs = require('@parcel/fs');

var os = require('os');

var IMPORT_RE = /^# *import +['"](.*)['"] *;? *$/;

var GraphqlAsset =
/*#__PURE__*/
function (_Asset) {
  _inherits(GraphqlAsset, _Asset);

  function GraphqlAsset(name, options) {
    var _this;

    _classCallCheck(this, GraphqlAsset);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(GraphqlAsset).call(this, name, options));
    _this.type = 'js';
    _this.gqlMap = new Map();
    _this.gqlResolver = new Resolver(Object.assign({}, _this.options, {
      extensions: ['.gql', '.graphql']
    }));
    return _this;
  }

  _createClass(GraphqlAsset, [{
    key: "traverseImports",
    value: function traverseImports(name, code) {
      var _this2 = this;

      return regeneratorRuntime.async(function traverseImports$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              this.gqlMap.set(name, code);
              _context2.next = 3;
              return regeneratorRuntime.awrap(Promise.all(code.split(/\r\n?|\n/).map(function (line) {
                return line.match(IMPORT_RE);
              }).filter(function (match) {
                return !!match;
              }).map(function _callee(_ref) {
                var _ref2, importName, _ref3, resolved, code;

                return regeneratorRuntime.async(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        _ref2 = _slicedToArray(_ref, 2), importName = _ref2[1];
                        _context.next = 3;
                        return regeneratorRuntime.awrap(_this2.gqlResolver.resolve(importName, name));

                      case 3:
                        _ref3 = _context.sent;
                        resolved = _ref3.path;

                        if (!_this2.gqlMap.has(resolved)) {
                          _context.next = 7;
                          break;
                        }

                        return _context.abrupt("return");

                      case 7:
                        _context.next = 9;
                        return regeneratorRuntime.awrap(fs.readFile(resolved, 'utf8'));

                      case 9:
                        code = _context.sent;
                        _context.next = 12;
                        return regeneratorRuntime.awrap(_this2.traverseImports(resolved, code));

                      case 12:
                      case "end":
                        return _context.stop();
                    }
                  }
                });
              })));

            case 3:
            case "end":
              return _context2.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "collectDependencies",
    value: function collectDependencies() {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this.gqlMap[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var _step$value = _slicedToArray(_step.value, 1),
              path = _step$value[0];

          this.addDependency(path, {
            includedInParent: true
          });
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }, {
    key: "parse",
    value: function parse(code) {
      var gql, allCodes;
      return regeneratorRuntime.async(function parse$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return regeneratorRuntime.awrap(localRequire('graphql-tag', this.name));

            case 2:
              gql = _context3.sent;
              _context3.next = 5;
              return regeneratorRuntime.awrap(this.traverseImports(this.name, code));

            case 5:
              allCodes = _toConsumableArray(this.gqlMap.values()).join(os.EOL);
              return _context3.abrupt("return", gql(allCodes));

            case 7:
            case "end":
              return _context3.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: "generate",
    value: function generate() {
      return "module.exports=".concat(JSON.stringify(this.ast, null, 2), ";");
    }
  }]);

  return GraphqlAsset;
}(Asset);

module.exports = GraphqlAsset;
