// jsdom/lib/jsdom/living/helpers/dist/selectors.dev.js
"use strict";

var DOMException = require("domexception");

var nwsapi = require("nwsapi");

var idlUtils = require("../generated/utils");

exports.matchesDontThrow = function (elImpl, selector) {
  var document = elImpl._ownerDocument;

  if (!document._nwsapiDontThrow) {
    document._nwsapiDontThrow = nwsapi({
      document: document,
      DOMException: DOMException
    });

    document._nwsapiDontThrow.configure({
      LOGERRORS: false,
      VERBOSITY: false,
      IDS_DUPES: true,
      MIXEDCASE: true
    });
  }

  return document._nwsapiDontThrow.match(selector, idlUtils.wrapperForImpl(elImpl));
}; // nwsapi gets `document.documentElement` at creation-time, so we have to initialize lazily, since in the initial
// stages of Document initialization, there is no documentElement present yet.


exports.addNwsapi = function (parentNode) {
  var document = parentNode._ownerDocument;

  if (!document._nwsapi) {
    document._nwsapi = nwsapi({
      document: document,
      DOMException: DOMException
    });

    document._nwsapi.configure({
      LOGERRORS: false,
      IDS_DUPES: true,
      MIXEDCASE: true
    });
  }

  return document._nwsapi;
};
