// parcel-bundler/lib/assets/dist/GLSLAsset.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var Asset = require('../Asset');

var localRequire = require('../utils/localRequire');

var path = require('path');

var _require = require('@parcel/utils'),
    promisify = _require.promisify;

var Resolver = require('../Resolver');

var GLSLAsset =
/*#__PURE__*/
function (_Asset) {
  _inherits(GLSLAsset, _Asset);

  function GLSLAsset(name, options) {
    var _this3;

    _classCallCheck(this, GLSLAsset);

    _this3 = _possibleConstructorReturn(this, _getPrototypeOf(GLSLAsset).call(this, name, options));
    _this3.type = 'js';
    return _this3;
  }

  _createClass(GLSLAsset, [{
    key: "parse",
    value: function parse() {
      var _this = this;

      return (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee2() {
        var glslifyDeps, resolver, cwd, depper;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return localRequire('glslify-deps', _this.name);

              case 2:
                glslifyDeps = _context2.sent;
                // Use the Parcel resolver rather than the default glslify one.
                // This adds support for parcel features like aliases, and tilde paths.
                resolver = new Resolver({
                  extensions: ['.glsl', '.vert', '.frag'],
                  rootDir: _this.options.rootDir
                }); // Parse and collect dependencies with glslify-deps

                cwd = path.dirname(_this.name);
                depper = glslifyDeps({
                  cwd: cwd,
                  resolve: function () {
                    var _resolve = (0, _asyncToGenerator2["default"])(
                    /*#__PURE__*/
                    regeneratorRuntime.mark(function _callee(target, opts, next) {
                      var res;
                      return regeneratorRuntime.wrap(function _callee$(_context) {
                        while (1) {
                          switch (_context.prev = _context.next) {
                            case 0:
                              _context.prev = 0;
                              _context.next = 3;
                              return resolver.resolve(target, path.join(opts.basedir, 'index'));

                            case 3:
                              res = _context.sent;
                              next(null, res.path);
                              _context.next = 10;
                              break;

                            case 7:
                              _context.prev = 7;
                              _context.t0 = _context["catch"](0);
                              next(_context.t0);

                            case 10:
                            case "end":
                              return _context.stop();
                          }
                        }
                      }, _callee, null, [[0, 7]]);
                    }));

                    function resolve(_x, _x2, _x3) {
                      return _resolve.apply(this, arguments);
                    }

                    return resolve;
                  }()
                });
                return _context2.abrupt("return", promisify(depper.inline.bind(depper))(_this.contents, cwd));

              case 7:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }))();
    }
  }, {
    key: "collectDependencies",
    value: function collectDependencies() {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this.ast[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var dep = _step.value;

          if (!dep.entry) {
            this.addDependency(dep.file, {
              includedInParent: true
            });
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }, {
    key: "generate",
    value: function generate() {
      var _this2 = this;

      return (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee3() {
        var glslifyBundle, glsl;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return localRequire('glslify-bundle', _this2.name);

              case 2:
                glslifyBundle = _context3.sent;
                glsl = glslifyBundle(_this2.ast);
                return _context3.abrupt("return", "module.exports=".concat(JSON.stringify(glsl), ";"));

              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3);
      }))();
    }
  }]);

  return GLSLAsset;
}(Asset);

module.exports = GLSLAsset;
