// parcel-bundler/lib/transforms/babel/dist/astConverter.dev.js
"use strict";

var traverse = require('@babel/traverse')["default"]; // Convert between babel 7 and babel 6 AST
// More info on the AST Changes: https://babeljs.io/docs/en/v7-migration-api#ast-changes


function babel7toBabel6(ast) {
  var visitor = {
    ArrowFunctionExpression: function ArrowFunctionExpression(node) {
      node.expression = node.body.type !== 'BlockStatement';
    },
    ExistsTypeAnnotation: function ExistsTypeAnnotation(node) {
      node.type = 'ExistentialTypeParam';
    },
    NumberLiteralTypeAnnotation: function NumberLiteralTypeAnnotation(node) {
      node.type = 'NumericLiteralTypeAnnotation';
    },
    ObjectTypeIndexer: function ObjectTypeIndexer(node) {
      node.end++;
      node.loc.end.column++;
    },
    ForOfStatement: function ForOfStatement(node) {
      node.type = 'ForAwaitStatement';
      delete node["await"];
    },
    SpreadElement: function SpreadElement(node, path) {
      if (path.parentPath.isObjectExpression() || path.parentPath.isArrayExpression()) {
        node.type = 'SpreadProperty';
      }
    },
    RestElement: function RestElement(node, path) {
      if (path.parentPath.isObjectPattern() || path.parentPath.isArrayPattern()) {
        node.type = 'RestProperty';
      }
    }
  };
  traverse(ast, {
    enter: function enter(path) {
      if (path.node.variance && path.node.variance.type === 'Variance') {
        path.node.variance = path.node.variance.kind;
      }

      var visitorFunc = visitor[path.node.type];

      if (visitorFunc) {
        visitorFunc(path.node, path);
      }
    }
  });
  return ast;
}

function babel6toBabel7(ast) {
  var visitor = {
    ArrowFunctionExpression: function ArrowFunctionExpression(node) {
      delete node.expression;
    },
    ExistentialTypeParam: function ExistentialTypeParam(node) {
      node.type = 'ExistsTypeAnnotation';
    },
    NumericLiteralTypeAnnotation: function NumericLiteralTypeAnnotation(node) {
      node.type = 'NumberLiteralTypeAnnotation';
    },
    ObjectTypeIndexer: function ObjectTypeIndexer(node) {
      node.end--;
      node.loc.end.column--;
    },
    ForAwaitStatement: function ForAwaitStatement(node) {
      node.type = 'ForOfStatement';
      node["await"] = true;
    },
    SpreadProperty: function SpreadProperty(node) {
      node.type = 'SpreadElement';
    },
    RestProperty: function RestProperty(node) {
      node.type = 'RestElement';
    }
  };
  traverse(ast, {
    enter: function enter(path) {
      if (path.node.variance && typeof path.node.variance === 'string') {
        path.node.variance = {
          type: 'VarianceNode',
          kind: path.node.variance
        };
      }

      var visitorFunc = visitor[path.node.type];

      if (visitorFunc) {
        visitorFunc(path.node);
      }
    }
  });
  return ast;
}

exports.babel7toBabel6 = babel7toBabel6;
exports.babel6toBabel7 = babel6toBabel7;
